--- ../src-base/minecraft/net/minecraftforge/event/terraingen/TerrainGen.java
+++ ../src-work/minecraft/net/minecraftforge/event/terraingen/TerrainGen.java
@@ -106,6 +106,7 @@
     {
         SaplingGrowTreeEvent event = new SaplingGrowTreeEvent(world, rand, pos);
         MinecraftForge.TERRAIN_GEN_BUS.post(event);
-        return event.getResult() != Result.DENY;
+        //Lava-test - add TreeType setter for bukkit event here TODO move to LavaEventHandler?
+        return event.getResult() != Result.DENY ? org.lavapowered.lava.events.LavaEventFactory.setTreeType(rand) : false;
     }
 }
