--- ../src-base/minecraft/net/minecraft/entity/projectile/EntitySmallFireball.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntitySmallFireball.java
@@ -21,6 +21,9 @@
     {
         super(worldIn, shooter, accelX, accelY, accelZ);
         this.setSize(0.3125F, 0.3125F);
+        if (this.shootingEntity != null && this.shootingEntity instanceof EntityLiving) {
+            isIncendiary = this.world.getGameRules().getBoolean("mobGriefing");
+        }
     }
 
     public EntitySmallFireball(World worldIn, double x, double y, double z, double accelX, double accelY, double accelZ)
@@ -43,11 +46,20 @@
                 if (!result.entityHit.isImmuneToFire())
                 {
                     boolean flag = result.entityHit.attackEntityFrom(DamageSource.causeFireballDamage(this, this.shootingEntity), 5.0F);
+                    isIncendiary = flag;
 
                     if (flag)
                     {
                         this.applyEnchantments(this.shootingEntity, result.entityHit);
-                        result.entityHit.setFire(5);
+                        //Craftbukkit start - add event
+                        // result.entityHit.setFire(5);
+                        org.bukkit.event.entity.EntityCombustByEntityEvent event = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), result.entityHit.getBukkitEntity(), 5);
+                        result.entityHit.world.getServer().getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled()) {
+                            result.entityHit.setFire(event.getDuration());
+                        }
+                        //Craftbukkit end
                     }
                 }
             }
@@ -60,13 +72,17 @@
                     flag1 = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity);
                 }
 
-                if (flag1)
+                if (flag1 && isIncendiary)
                 {
                     BlockPos blockpos = result.getBlockPos().offset(result.sideHit);
 
                     if (this.world.isAirBlock(blockpos))
                     {
-                        this.world.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
+                        //Craftbukkit start - event
+                        //this.world.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos.getX(), blockpos.getY(), blockpos.getZ(), this).isCancelled()) {
+                            this.world.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
+                        }
                     }
                 }
             }
