--- ../src-base/minecraft/net/minecraft/entity/passive/EntityMooshroom.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityMooshroom.java
@@ -41,8 +41,7 @@
             if (itemstack.isEmpty())
             {
                 player.setHeldItem(hand, new ItemStack(Items.MUSHROOM_STEW));
-            }
-            else if (!player.inventory.addItemStackToInventory(new ItemStack(Items.MUSHROOM_STEW)))
+            } else if (!player.inventory.addItemStackToInventory(new ItemStack(Items.MUSHROOM_STEW)))
             {
                 player.dropItem(new ItemStack(Items.MUSHROOM_STEW), false);
             }
@@ -91,6 +90,7 @@
     }
 
     @Override public boolean isShearable(ItemStack item, net.minecraft.world.IBlockAccess world, net.minecraft.util.math.BlockPos pos){ return getGrowingAge() >= 0; }
+
     @Override
     public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, net.minecraft.util.math.BlockPos pos, int fortune)
     {
