--- ../src-base/minecraft/net/minecraft/entity/passive/EntityPig.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityPig.java
@@ -188,7 +188,7 @@
 
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+        // super.onDeath(cause); // CraftBukkit - Moved to end
 
         if (!this.world.isRemote)
         {
@@ -197,6 +197,7 @@
                 this.dropItem(Items.SADDLE, 1);
             }
         }
+        super.onDeath(cause); // CraftBukkit - Moved from above
     }
 
     @Nullable
@@ -222,11 +223,22 @@
         }
     }
 
-    public void onStruckByLightning(EntityLightningBolt lightningBolt)
+    public void onStruckByLightning(@Nullable EntityLightningBolt lightningBolt) //Craftbukkit - add @Nullable
     {
         if (!this.world.isRemote && !this.isDead)
         {
+            //Craftbukkit start
+            if (lightningBolt == null)
+                lightningBolt = new EntityLightningBolt(this.world, this.posX, this.posY, this.posZ, false);
+            //Craftbukkit end
             EntityPigZombie entitypigzombie = new EntityPigZombie(this.world);
+
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callPigZapEvent(this, lightningBolt, entitypigzombie).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
             entitypigzombie.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
             entitypigzombie.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
             entitypigzombie.setNoAI(this.isAIDisabled());
@@ -237,7 +249,10 @@
                 entitypigzombie.setAlwaysRenderNameTag(this.getAlwaysRenderNameTag());
             }
 
-            this.world.spawnEntity(entitypigzombie);
+            // CraftBukkit start - added a reason for spawning this creature
+            //this.world.spawnEntity(entitypigzombie);
+            this.world.spawnEntity(entitypigzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+            //Craftbukkit end
             this.setDead();
         }
     }
