--- ../src-base/minecraft/net/minecraft/entity/passive/EntityWolf.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityWolf.java
@@ -395,7 +395,10 @@
                             itemstack.shrink(1);
                         }
 
-                        this.heal((float)itemfood.getHealAmount(itemstack));
+                        //Craftbukkit start - add regain reason
+                        //this.heal((float)itemfood.getHealAmount(itemstack));
+                        this.heal((float)itemfood.getHealAmount(itemstack), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING);
+                        //Craftbukkit end
                         return true;
                     }
                 }
@@ -422,7 +425,10 @@
                 this.aiSit.setSitting(!this.isSitting());
                 this.isJumping = false;
                 this.navigator.clearPath();
-                this.setAttackTarget((EntityLivingBase)null);
+                //Craftbukkit start - add target reason
+                //this.setAttackTarget((EntityLivingBase)null);
+                this.setAttackTarget((EntityLivingBase)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true);
+                //Craftbukkit end
             }
         }
         else if (itemstack.getItem() == Items.BONE && !this.isAngry())
@@ -440,7 +446,10 @@
                     this.navigator.clearPath();
                     this.setAttackTarget((EntityLivingBase)null);
                     this.aiSit.setSitting(true);
-                    this.setHealth(20.0F);
+                    // CraftBukkit start - 20.0 -> getMaxHealth()
+                    // this.setHealth(20.0F);
+                    this.setHealth(this.getMaxHealth());
+                    //Craftbukkit end
                     this.playTameEffect(true);
                     this.world.setEntityState(this, (byte)7);
                 }
