--- ../src-base/minecraft/net/minecraft/entity/passive/EntityCow.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityCow.java
@@ -95,16 +95,26 @@
 
         if (itemstack.getItem() == Items.BUCKET && !player.capabilities.isCreativeMode && !this.isChild())
         {
+            //Craftbukkit start - event
+            org.bukkit.Location loc = this.getBukkitEntity().getLocation();
+            org.bukkit.event.player.PlayerBucketFillEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketFillEvent(player, loc.getBlockX(), loc.getBlockY(), loc.getBlockZ(), null, itemstack, Items.MILK_BUCKET);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+            ItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack());
+            //Craftbukkit end
+
             player.playSound(SoundEvents.ENTITY_COW_MILK, 1.0F, 1.0F);
             itemstack.shrink(1);
 
             if (itemstack.isEmpty())
             {
-                player.setHeldItem(hand, new ItemStack(Items.MILK_BUCKET));
+                player.setHeldItem(hand, result); //Craftbukkit new ItemStack(Items.MILK_BUCKET) -> result
             }
-            else if (!player.inventory.addItemStackToInventory(new ItemStack(Items.MILK_BUCKET)))
+            else if (!player.inventory.addItemStackToInventory(result)) //Craftbukkit new ItemStack(Items.MILK_BUCKET) -> result
             {
-                player.dropItem(new ItemStack(Items.MILK_BUCKET), false);
+                player.dropItem(result, false); //Craftbukkit new ItemStack(Items.MILK_BUCKET) -> result
             }
 
             return true;
