--- ../src-base/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
@@ -33,15 +33,15 @@
     private static final DataParameter<Integer> PARTICLE_PARAM_1 = EntityDataManager.<Integer>createKey(EntityAreaEffectCloud.class, DataSerializers.VARINT);
     private static final DataParameter<Integer> PARTICLE_PARAM_2 = EntityDataManager.<Integer>createKey(EntityAreaEffectCloud.class, DataSerializers.VARINT);
     private PotionType potion;
-    private final List<PotionEffect> effects;
+    public final List<PotionEffect> effects; //Craftbukkit - private -> public
     private final Map<Entity, Integer> reapplicationDelayMap;
     private int duration;
-    private int waitTime;
-    private int reapplicationDelay;
+    public int waitTime; //Craftbukkit - private -> public
+    public int reapplicationDelay; //Craftbukkit - private -> public
     private boolean colorSet;
-    private int durationOnUse;
-    private float radiusOnUse;
-    private float radiusPerTick;
+    public int durationOnUse; //Craftbukkit - private -> public
+    public float radiusOnUse; //Craftbukkit - private -> public
+    public float radiusPerTick; //Craftbukkit - private -> public
     private EntityLivingBase owner;
     private UUID ownerUniqueId;
 
@@ -126,6 +126,23 @@
         }
     }
 
+    // CraftBukkit start accessor methods
+    public void refreshEffects() {
+        if (!this.colorSet) {
+            this.getDataManager().set(EntityAreaEffectCloud.COLOR, PotionUtils.getPotionColorFromEffectList(PotionUtils.mergeEffects(this.potion, (java.util.Collection) this.effects)));
+        }
+    }
+
+    public String getType() {
+        return (PotionType.REGISTRY.getNameForObject(this.potion)).toString();
+    }
+
+    public void setType(String string) {
+        setPotion(PotionType.REGISTRY.getObject(new ResourceLocation(string)));
+    }
+    // CraftBukkit end
+
+
     public int getColor()
     {
         return ((Integer)this.getDataManager().get(COLOR)).intValue();
@@ -326,6 +343,7 @@
 
                     if (!list.isEmpty())
                     {
+                        List<org.bukkit.entity.LivingEntity> entities = new java.util.ArrayList<>(); //Craftbukkit store entities for event
                         for (EntityLivingBase entitylivingbase : list)
                         {
                             if (!this.reapplicationDelayMap.containsKey(entitylivingbase) && entitylivingbase.canBeHitWithPotion())
@@ -336,6 +354,17 @@
 
                                 if (d2 <= (double)(f * f))
                                 {
+                                    //Craftbukkit start - event
+                                    entities.add((org.bukkit.entity.LivingEntity) entitylivingbase.getBukkitEntity());
+                                }
+                            }
+                        }
+                        org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, entities);
+                        { // Preserve NMS spacing and bracket count for smallest diff
+                            for (org.bukkit.entity.LivingEntity entity : event.getAffectedEntities()) {
+                                if (entity instanceof org.bukkit.craftbukkit.entity.CraftLivingEntity) {
+                                    EntityLivingBase entitylivingbase = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) entity).getHandle();
+                                    //Craftbukkit end
                                     this.reapplicationDelayMap.put(entitylivingbase, Integer.valueOf(this.ticksExisted + this.reapplicationDelay));
 
                                     for (PotionEffect potioneffect : potions)
