--- ../src-base/minecraft/net/minecraft/entity/EntityLeashKnot.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLeashKnot.java
@@ -105,6 +105,12 @@
             {
                 if (entityliving.getLeashed() && entityliving.getLeashHolder() == player)
                 {
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerLeashEntityEvent(entityliving, this, player).isCancelled()) {
+                        ((net.minecraft.entity.player.EntityPlayerMP) player).connection.sendPacket(new net.minecraft.network.play.server.SPacketEntityAttach(entityliving, entityliving.getLeashHolder()));
+                        continue;
+                    }
+                    // CraftBukkit end
                     entityliving.setLeashHolder(this, true);
                     flag = true;
                 }
@@ -112,17 +118,28 @@
 
             if (!flag)
             {
-                this.setDead();
-
-                if (player.capabilities.isCreativeMode)
+                // CraftBukkit start - Move below
+                //this.setDead();
+                if (true/*player.capabilities.isCreativeMode*/) // CraftBukkit - Process for non-creative as well
                 {
+                    boolean die = true;
+                    //Craftbukkit end
                     for (EntityLiving entityliving1 : list)
                     {
+                        //Craftbukkit start - add event
                         if (entityliving1.getLeashed() && entityliving1.getLeashHolder() == this)
                         {
-                            entityliving1.clearLeashed(true, false);
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerUnleashEntityEvent(entityliving1, player).isCancelled()) {
+                                die = false;
+                                continue;
+                            }
+                            entityliving1.clearLeashed(true, !player.capabilities.isCreativeMode); // false -> survival mode boolean
                         }
                     }
+                    if (die) {
+                        this.setDead();
+                    }
+                    //Craftbukkit end
                 }
             }
 
