--- ../src-base/minecraft/net/minecraft/entity/EntityCreature.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityCreature.java
@@ -93,6 +93,7 @@
             {
                 if (f > 10.0F)
                 {
+                    this.world.getServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE)); //Craftbukkit - event
                     this.clearLeashed(true, true);
                 }
 
@@ -103,6 +104,7 @@
 
             if (f > 10.0F)
             {
+                this.world.getServer().getPluginManager().callEvent(new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE)); //Craftbukkit - event
                 this.clearLeashed(true, true);
                 this.tasks.disableControlFlag(1);
             }
