--- ../src-base/minecraft/net/minecraftforge/common/util/EnumHelper.java
+++ ../src-work/minecraft/net/minecraftforge/common/util/EnumHelper.java
@@ -428,4 +428,25 @@
             setup();
         }
     }
+    //Lava start - add bukkit enums TODO what is this for
+    public static org.bukkit.block.Biome addBukkitBiome(String name) {
+        return addEnum(org.bukkit.block.Biome.class, name, new Class[]{}, new Object[]{});
+    }
+
+    public static org.bukkit.World.Environment addBukkitEnvironment(int id, String name) {
+        if (!isSetup) {
+            setup();
+        }
+        return addEnum(org.bukkit.World.Environment.class, name, new Class[]{Integer.TYPE}, id);
+    }
+
+    public static org.bukkit.WorldType addBukkitWorldType(String name) {
+        if (!isSetup) {
+            setup();
+        }
+        org.bukkit.WorldType worldType = addEnum(org.bukkit.WorldType.class, name, new Class[]{String.class}, name);
+        Map<String, org.bukkit.WorldType> BY_NAME = net.minecraftforge.fml.relauncher.ReflectionHelper.getPrivateValue(org.bukkit.WorldType.class, null, "BY_NAME");
+        BY_NAME.put(name.toUpperCase(), worldType);
+        return worldType;
+    }
 }
