--- ../src-base/minecraft/net/minecraft/inventory/ContainerMerchant.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerMerchant.java
@@ -12,6 +12,11 @@
     private final InventoryMerchant merchantInventory;
     private final World world;
 
+    //Craftbukkit start - fields
+    private org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    //Craftbukkit end
+
     public ContainerMerchant(InventoryPlayer playerInventory, IMerchant merchant, World worldIn)
     {
         this.merchant = merchant;
@@ -20,6 +25,7 @@
         this.addSlotToContainer(new Slot(this.merchantInventory, 0, 36, 53));
         this.addSlotToContainer(new Slot(this.merchantInventory, 1, 62, 53));
         this.addSlotToContainer(new SlotMerchantResult(playerInventory.player, merchant, this.merchantInventory, 2, 120, 53));
+        this.player = playerInventory; //Craftbukkit
 
         for (int i = 0; i < 3; ++i)
         {
@@ -137,4 +143,13 @@
             }
         }
     }
+
+    //Craftbukkit method
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity == null) {
+            bukkitEntity = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.player.player.getBukkitEntity(), new org.bukkit.craftbukkit.inventory.CraftInventoryMerchant(merchantInventory), this);
+        }
+        return bukkitEntity;
+    }
 }
