--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseChest.java
@@ -6,6 +6,19 @@
 
 public class ContainerHorseChest extends InventoryBasic
 {
+    //Craftbukkit start - field, method and constructor ///Lava-test - keep old constructor for compatibility
+    private net.minecraft.entity.passive.EntityAnimal entityAnimal;
+
+    public ContainerHorseChest(String inventoryTitle, int slotCount, net.minecraft.entity.passive.AbstractHorse owner) {
+        super(inventoryTitle, false, slotCount, (org.bukkit.entity.AbstractHorse) owner.getBukkitEntity());
+        this.entityAnimal = owner;
+    }
+
+    public net.minecraft.entity.passive.EntityAnimal getAnimal() {
+        return this.entityAnimal;
+    }
+    //Craftbukkit end
+
     public ContainerHorseChest(String inventoryTitle, int slotCount)
     {
         super(inventoryTitle, false, slotCount);
