--- ../src-base/minecraft/net/minecraft/inventory/ContainerBeacon.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerBeacon.java
@@ -12,8 +12,14 @@
     private final IInventory tileBeacon;
     private final ContainerBeacon.BeaconSlot beaconSlot;
 
+    //Craftbukkit start - fields
+    private org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity = null;
+    private net.minecraft.entity.player.InventoryPlayer player;
+    //Craftbukkit end
+
     public ContainerBeacon(IInventory playerInventory, IInventory tileBeaconIn)
     {
+        player = (net.minecraft.entity.player.InventoryPlayer) playerInventory; // CraftBukkit
         this.tileBeacon = tileBeaconIn;
         this.beaconSlot = new ContainerBeacon.BeaconSlot(tileBeaconIn, 0, 136, 110);
         this.addSlotToContainer(this.beaconSlot);
@@ -68,6 +74,7 @@
 
     public boolean canInteractWith(EntityPlayer playerIn)
     {
+        if (!this.checkReachable) return true; //Craftbukkit - overrides
         return this.tileBeacon.isUsableByPlayer(playerIn);
     }
 
@@ -150,4 +157,16 @@
             return 1;
         }
     }
+
+    //Craftbukkit method
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryBeacon((net.minecraft.tileentity.TileEntityBeacon) this.tileBeacon);
+        bukkitEntity = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
 }
