--- ../src-base/minecraft/net/minecraft/inventory/ContainerWorkbench.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerWorkbench.java
@@ -15,8 +15,17 @@
     private final BlockPos pos;
     private final EntityPlayer player;
 
+    
+    private org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer playerInventory;
+
     public ContainerWorkbench(InventoryPlayer playerInventory, World worldIn, BlockPos posIn)
     {
+        // CraftBukkit start
+        this.craftMatrix.owner = playerInventory.player;
+        this.craftMatrix.resultInventory = this.craftResult;
+        this.playerInventory = playerInventory;
+        // CraftBukkit end
         this.world = worldIn;
         this.pos = posIn;
         this.player = playerInventory.player;
@@ -61,6 +70,7 @@
 
     public boolean canInteractWith(EntityPlayer playerIn)
     {
+        if (!this.checkReachable) return true; //Craftbukkit - override for virtual anvils
         if (this.world.getBlockState(this.pos).getBlock() != Blocks.CRAFTING_TABLE)
         {
             return false;
@@ -140,4 +150,16 @@
     {
         return slotIn.inventory != this.craftResult && super.canMergeSlot(stack, slotIn);
     }
+
+    //Craftbukkit method
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventoryCrafting inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryCrafting(this.craftMatrix, this.craftResult);
+        bukkitEntity = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.playerInventory.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
 }
