--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
@@ -7,14 +7,22 @@
 import net.minecraft.item.ItemStack;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class ContainerHorseInventory extends Container
 {
     private final IInventory horseInventory;
     private final AbstractHorse horse;
 
+    //Craftbukkit start - fields
+    org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity;
+    net.minecraft.entity.player.InventoryPlayer player;
+    //Craftbukkit end
+
     public ContainerHorseInventory(IInventory playerInventory, IInventory horseInventoryIn, final AbstractHorse horse, EntityPlayer player)
     {
+        this.player = (net.minecraft.entity.player.InventoryPlayer) playerInventory; //Craftbukkit
         this.horseInventory = horseInventoryIn;
         this.horse = horse;
         int i = 3;
@@ -133,4 +141,9 @@
         super.onContainerClosed(playerIn);
         this.horseInventory.closeInventory(playerIn);
     }
+
+    //Craftbukkit method
+    @Override
+    public InventoryView getBukkitView() { return bukkitEntity == null ? bukkitEntity = new CraftInventoryView(player.player.getBukkitEntity(), horseInventory.getOwner().getInventory(), this) : bukkitEntity; }
+
 }
