--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseInventory.java
@@ -13,8 +13,14 @@
     private final IInventory horseInventory;
     private final AbstractHorse horse;
 
+    //Craftbukkit start - fields
+    org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity;
+    net.minecraft.entity.player.InventoryPlayer player;
+    //Craftbukkit end
+
     public ContainerHorseInventory(IInventory playerInventory, IInventory horseInventoryIn, final AbstractHorse horse, EntityPlayer player)
     {
+        this.player = (net.minecraft.entity.player.InventoryPlayer) playerInventory; //Craftbukkit
         this.horseInventory = horseInventoryIn;
         this.horse = horse;
         int i = 3;
@@ -133,4 +139,9 @@
         super.onContainerClosed(playerIn);
         this.horseInventory.closeInventory(playerIn);
     }
+
+    //Craftbukkit method
+    @Override
+    public org.bukkit.inventory.InventoryView getBukkitView() { return bukkitEntity == null ? bukkitEntity = new org.bukkit.craftbukkit.inventory.CraftInventoryView(player.player.getBukkitEntity(), horseInventory.getOwner().getInventory(), this) : bukkitEntity; }
+
 }
