--- ../src-base/minecraft/net/minecraft/inventory/InventoryLargeChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/InventoryLargeChest.java
@@ -12,9 +12,51 @@
 public class InventoryLargeChest implements ILockableContainer
 {
     private final String name;
-    private final ILockableContainer upperChest;
-    private final ILockableContainer lowerChest;
+    public final ILockableContainer upperChest; //Craftbukkit - private -> public
+    public final ILockableContainer lowerChest; //Craftbukkit - private -> public
 
+    //Craftbukkit start - methods..
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+
+    public java.util.List<ItemStack> getContents() {
+        java.util.List<ItemStack> result = new java.util.ArrayList<ItemStack>(this.getSizeInventory());
+        for (int i = 0; i < this.getSizeInventory(); i++) {
+            result.add(this.getStackInSlot(i));
+        }
+        return result;
+    }
+
+    public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+        this.upperChest.onOpen(who);
+        this.lowerChest.onOpen(who);
+        transaction.add(who);
+    }
+
+    public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+         this.upperChest.onClose(who);
+         this.lowerChest.onClose(who);
+        transaction.remove(who);
+    }
+
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public org.bukkit.inventory.InventoryHolder getOwner() {
+        return null; // This method won't be called since CraftInventoryDoubleChest doesn't defer to here
+    }
+
+    public void setMaxStackSize(int size) {
+        this.upperChest.setMaxStackSize(size);
+        this.lowerChest.setMaxStackSize(size);
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        return upperChest.getLocation(); // TODO: right?
+    }
+    //Craftbukkit end
+
     public InventoryLargeChest(String nameIn, ILockableContainer upperChestIn, ILockableContainer lowerChestIn)
     {
         this.name = nameIn;
@@ -108,7 +150,10 @@
 
     public int getInventoryStackLimit()
     {
-        return this.upperChest.getInventoryStackLimit();
+        //Craftbukkit start
+        // return this.upperChest.getInventoryStackLimit();
+        return Math.min(this.upperChest.getInventoryStackLimit(), this.lowerChest.getInventoryStackLimit()); //check both sides
+        //Craftbukkit end
     }
 
     public void markDirty()
