--- ../src-base/minecraft/net/minecraft/inventory/InventoryCraftResult.java
+++ ../src-work/minecraft/net/minecraft/inventory/InventoryCraftResult.java
@@ -14,6 +14,39 @@
     private final NonNullList<ItemStack> stackResult = NonNullList.<ItemStack>withSize(1, ItemStack.EMPTY);
     private IRecipe recipeUsed;
 
+    //Craftbukkit start - fields, methods..
+    private int maxStack = MAX_STACK;
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+
+    public void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return transaction;
+    }
+    public java.util.List<ItemStack> getContents() {
+        return this.stackResult;
+    }
+
+    public org.bukkit.inventory.InventoryHolder getOwner() {
+        return null; // Result slots don't get an owner
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        return null;
+    }
+    //Craftbukkit end
+
     public int getSizeInventory()
     {
         return 1;
@@ -69,7 +102,7 @@
 
     public int getInventoryStackLimit()
     {
-        return 64;
+        return maxStack; //Craftbukkit
     }
 
     public void markDirty()
