--- ../src-base/minecraft/net/minecraft/inventory/InventoryEnderChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/InventoryEnderChest.java
@@ -10,11 +10,21 @@
 {
     private TileEntityEnderChest associatedChest;
 
-    public InventoryEnderChest()
-    {
+    //Craftbukkit start
+    public EntityPlayer owner; //Lava-test - private -> public -final
+
+    public InventoryEnderChest(EntityPlayer owner) {
         super("container.enderchest", false, 27);
+        this.owner = owner;
     }
+    //Craftbukkit end
 
+    public InventoryEnderChest(){
+        super("container.enderchest", false, 27);
+        this.owner = null; //Craftbukkit
+    }
+
+
     public void setChestTileEntity(TileEntityEnderChest chestTileEntity)
     {
         this.associatedChest = chestTileEntity;
@@ -84,4 +94,15 @@
         super.closeInventory(player);
         this.associatedChest = null;
     }
+
+    //Craftbukkit methods
+    public org.bukkit.inventory.InventoryHolder getBukkitOwner() {
+        return owner.getBukkitEntity();
+    }
+
+    @Override
+    public org.bukkit.Location getLocation() {
+        return new org.bukkit.Location(this.associatedChest.getWorld().getWorld(), this.associatedChest.getPos().getX(), this.associatedChest.getPos().getY(), this.associatedChest.getPos().getZ());
+    }
+
 }
