--- ../src-base/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
+++ ../src-work/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -70,11 +70,26 @@
                 i = j;
             }
 
+            //Craftbukkit start - event
+            org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.player.getBukkitEntity();
+            net.minecraft.tileentity.TileEntityFurnace furnace = ((net.minecraft.tileentity.TileEntityFurnace) this.inventory);
+            org.bukkit.block.Block block = this.player.world.getWorld().getBlockAt(furnace.getPos().getX(), furnace.getPos().getY(), furnace.getPos().getZ());
+
+            if (removeCount != 0) {
+                org.bukkit.event.inventory.FurnaceExtractEvent event = new org.bukkit.event.inventory.FurnaceExtractEvent(player, block, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.getItem()), removeCount, i);
+                this.player.world.getServer().getPluginManager().callEvent(event);
+                i = event.getExpToDrop();
+            }
+            //Craftbukkit end
+
             while (i > 0)
             {
                 int k = EntityXPOrb.getXPSplit(i);
                 i -= k;
-                this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k));
+                //Craftbukkit start - add spawn reason and trigger entity
+                //this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k));
+                this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k, org.bukkit.entity.ExperienceOrb.SpawnReason.FURNACE, this.player));
+                //Craftbukkit end
             }
         }
 
