--- ../src-base/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
@@ -21,6 +21,10 @@
     {
         return "Rcon";
     }
+    //Craftbukkit method? - TODO not used anywhere
+    public void sendMessage(String message) {
+        this.buffer.append(message);
+    }
 
     public void sendMessage(ITextComponent component)
     {
