--- ../src-base/minecraft/net/minecraftforge/event/ForgeEventFactory.java
+++ ../src-work/minecraft/net/minecraftforge/event/ForgeEventFactory.java
@@ -34,6 +34,7 @@
 import net.minecraft.entity.IMerchant;
 import net.minecraft.entity.effect.EntityLightningBolt;
 import net.minecraft.entity.item.EntityItem;
+import net.minecraft.entity.monster.EntityEnderman;
 import net.minecraft.entity.monster.EntityZombie;
 import net.minecraft.entity.passive.EntityAnimal;
 import net.minecraft.entity.player.EntityPlayer;
@@ -141,6 +142,9 @@
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.lavapowered.lava.events.LavaEventFactory;
+
 public class ForgeEventFactory
 {
 
@@ -167,6 +171,10 @@
         IBlockState placedAgainst = blockSnapshot.getWorld().getBlockState(blockSnapshot.getPos().offset(direction.getOpposite()));
         EntityPlaceEvent event = new BlockEvent.EntityPlaceEvent(blockSnapshot, placedAgainst, entity);
         MinecraftForge.EVENT_BUS.post(event);
+        //Lava-test - move bukkit event here
+        if (entity instanceof EntityEnderman) {
+            event.setCanceled(org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockSnapshot.getPos(), ((EntityEnderman)entity).getHeldBlockState().getBlock(), ((EntityEnderman)entity).getHeldBlockState().getBlock().getMetaFromState(((EntityEnderman)entity).getHeldBlockState())).isCancelled());
+        }
         return event;
     }
 
@@ -342,13 +350,24 @@
     {
         BlockEvent.HarvestDropsEvent event = new BlockEvent.HarvestDropsEvent(world, pos, state, fortune, dropChance, drops, player, silkTouch);
         MinecraftForge.EVENT_BUS.post(event);
-        return event.getDropChance();
+        //Lava-test - handle Craftbukkit's change here
+        return event.getDropChance() == 0.0f ? -1.0f : event.getDropChance();
     }
 
     public static IBlockState fireFluidPlaceBlockEvent(World world, BlockPos pos, BlockPos liquidPos, IBlockState state)
     {
         BlockEvent.FluidPlaceBlockEvent event = new BlockEvent.FluidPlaceBlockEvent(world, pos, liquidPos, state);
         MinecraftForge.EVENT_BUS.post(event);
+        
+        //Lava-test - move bukkit events here
+        if (state == Blocks.FIRE.getDefaultState()) { //TODO pass the changed state or the original one? (the formed block)
+        	return !CraftEventFactory.callBlockIgniteEvent(world, pos.getX(), pos.getY(), pos.getZ(), liquidPos.getX(), liquidPos.getY(), liquidPos.getZ()).isCancelled() ? event.getNewState() : event.getOriginalState();
+        }
+        
+        else if ((state == Blocks.COBBLESTONE.getDefaultState()) || (state == Blocks.STONE.getDefaultState()) || (state == Blocks.OBSIDIAN.getDefaultState())) {
+        	return CraftEventFactory.handleBlockFormEvent(world, pos, event.getNewState(), null) ? event.getNewState() : event.getOriginalState();
+        }
+        
         return event.getNewState();
     }
 
@@ -503,7 +522,7 @@
     {
         if (item.isEmpty()) return -1;
         ItemExpireEvent event = new ItemExpireEvent(entity, (item.isEmpty() ? 6000 : item.getItem().getEntityLifespan(item, entity.world)));
-        if (!MinecraftForge.EVENT_BUS.post(event)) return -1;
+        if (!MinecraftForge.EVENT_BUS.post(event) && CraftEventFactory.callItemDespawnEvent(entity).isCancelled()) return -1; //Lava-test - move bukkit event here
         return event.getExtraLife();
     }
 
@@ -511,7 +530,7 @@
     {
         Event event = new EntityItemPickupEvent(player, entityItem);
         if (MinecraftForge.EVENT_BUS.post(event)) return -1;
-        return event.getResult() == Result.ALLOW ? 1 : 0;
+        return event.getResult() == Result.ALLOW ? LavaEventFactory.handleItemPickupEvent(entityItem, player) : 0;
     }
 
     public static void onPlayerDrops(EntityPlayer player, DamageSource cause, List<EntityItem> capturedDrops, boolean recentlyHit)
@@ -541,7 +560,8 @@
 
     public static boolean onAnimalTame(EntityAnimal animal, EntityPlayer tamer)
     {
-        return MinecraftForge.EVENT_BUS.post(new AnimalTameEvent(animal, tamer));
+        //Lava-test - moved tame event here
+        return MinecraftForge.EVENT_BUS.post(new AnimalTameEvent(animal, tamer)) && !CraftEventFactory.callEntityTameEvent(animal, tamer).isCancelled();
     }
 
     public static SleepResult onPlayerSleepInBed(EntityPlayer player, BlockPos pos)
@@ -788,7 +808,8 @@
 
     public static boolean onTrySpawnPortal(World world, BlockPos pos, BlockPortal.Size size)
     {
-        return MinecraftForge.EVENT_BUS.post(new BlockEvent.PortalSpawnEvent(world, pos, world.getBlockState(pos), size));
+        //Lava-test - move bukkit handling here
+        return MinecraftForge.EVENT_BUS.post(new BlockEvent.PortalSpawnEvent(world, pos, world.getBlockState(pos), size)) ? size.placePortalBlocksCB() : false;
     }
 
     public static int onEnchantmentLevelSet(World world, BlockPos pos, int enchantRow, int power, ItemStack itemStack, int level)
