--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIHarvestFarmland.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIHarvestFarmland.java
@@ -61,6 +61,13 @@
 
             if (this.currentTask == 0 && block instanceof BlockCrops && ((BlockCrops)block).isMaxAge(iblockstate))
             {
+                //Lava-test start - rework event to preserve original code block
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, Blocks.AIR, 0).isCancelled()) {
+                    this.currentTask = -1; //add rest of code TODO test if this doesn't interfere with coremods
+                    this.runDelay = 10;
+                    return;
+                }
+                //Lava-test end
                 world.destroyBlock(blockpos, true);
             }
             else if (this.currentTask == 1 && iblockstate.getMaterial() == Material.AIR)
@@ -76,27 +83,101 @@
                     {
                         if (itemstack.getItem() == Items.WHEAT_SEEDS)
                         {
+                            //Lava-test start - add event and preserve original code block
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, Blocks.WHEAT, 0).isCancelled()) {
+                                if (flag) { //TODO should I create a method for this?
+                                    itemstack.shrink(1);
+
+                                    if (itemstack.isEmpty()) {
+                                        inventorybasic.setInventorySlotContents(i, ItemStack.EMPTY);
+                                    }
+
+                                    break;
+                                }
+                                
+                                return;
+                            }
+                            //Lava-test end
                             world.setBlockState(blockpos, Blocks.WHEAT.getDefaultState(), 3);
                             flag = true;
                         }
                         else if (itemstack.getItem() == Items.POTATO)
                         {
+                          //Lava-test start - add event and preserve original code block
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, Blocks.POTATOES, 0).isCancelled()) {
+                                if (flag) { //TODO should I create a method for this?
+                                    itemstack.shrink(1);
+
+                                    if (itemstack.isEmpty()) {
+                                        inventorybasic.setInventorySlotContents(i, ItemStack.EMPTY);
+                                    }
+
+                                    break;
+                                }
+                                
+                                return;
+                            }
+                            //Lava-test end
                             world.setBlockState(blockpos, Blocks.POTATOES.getDefaultState(), 3);
                             flag = true;
                         }
                         else if (itemstack.getItem() == Items.CARROT)
                         {
+                          //Lava-test start - add event and preserve original code block
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, Blocks.CARROTS, 0).isCancelled()) {
+                                if (flag) { //TODO should I create a method for this?
+                                    itemstack.shrink(1);
+
+                                    if (itemstack.isEmpty()) {
+                                        inventorybasic.setInventorySlotContents(i, ItemStack.EMPTY);
+                                    }
+
+                                    break;
+                                }
+                                
+                                return;
+                            }
+                            //Lava-test end
                             world.setBlockState(blockpos, Blocks.CARROTS.getDefaultState(), 3);
                             flag = true;
                         }
                         else if (itemstack.getItem() == Items.BEETROOT_SEEDS)
                         {
+                          //Lava-test start - add event and preserve original code block
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, Blocks.BEETROOTS, 0).isCancelled()) {
+                                if (flag) { //TODO should I create a method for this?
+                                    itemstack.shrink(1);
+
+                                    if (itemstack.isEmpty()) {
+                                        inventorybasic.setInventorySlotContents(i, ItemStack.EMPTY);
+                                    }
+
+                                    break;
+                                }
+                                
+                                return;
+                            }
+                            //Lava-test end
                             world.setBlockState(blockpos, Blocks.BEETROOTS.getDefaultState(), 3);
                             flag = true;
-                        }
-                        else if (itemstack.getItem() instanceof net.minecraftforge.common.IPlantable) {
-                            if(((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlantType(world,blockpos) == net.minecraftforge.common.EnumPlantType.Crop) {
-                                world.setBlockState(blockpos, ((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlant(world,blockpos),3);
+                        } else if (itemstack.getItem() instanceof net.minecraftforge.common.IPlantable) {
+                            if (((net.minecraftforge.common.IPlantable) itemstack.getItem()).getPlantType(world, blockpos) == net.minecraftforge.common.EnumPlantType.Crop) {
+                              //Lava-test start - add event and preserve original code block
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.villager, blockpos, ((net.minecraftforge.common.IPlantable) itemstack.getItem()).getPlant(world, blockpos).getBlock(), 0).isCancelled()) {
+                                    if (flag) { //TODO should I create a method for this?
+                                        itemstack.shrink(1);
+
+                                        if (itemstack.isEmpty()) {
+                                            inventorybasic.setInventorySlotContents(i, ItemStack.EMPTY);
+                                        }
+
+                                        break;
+                                    }
+                                    
+                                    return;
+                                }
+                                //Lava-test end
+                                world.setBlockState(blockpos, ((net.minecraftforge.common.IPlantable) itemstack.getItem()).getPlant(world, blockpos), 3);
                                 flag = true;
                             }
                         }
