--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIEatGrass.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIEatGrass.java
@@ -10,6 +10,8 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityAIEatGrass extends EntityAIBase
 {
@@ -80,6 +82,12 @@
             {
                 if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                 {
+                    //Lava-test start - implement event like this to preserve original code block
+                    if (CraftEventFactory.callEntityChangeBlockEvent(this.grassEaterEntity, this.grassEaterEntity.world.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()), Material.AIR, false).isCancelled()) {
+                        this.grassEaterEntity.eatGrassBonus();
+                        return; //skip the destroyBlock call
+                    }
+                    //Lava-test end
                     this.entityWorld.destroyBlock(blockpos, false);
                 }
 
@@ -93,6 +101,12 @@
                 {
                     if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                     {
+                        //Lava-test start - event
+                        if (CraftEventFactory.callEntityChangeBlockEvent(this.grassEaterEntity, this.grassEaterEntity.world.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()), Material.AIR, false).isCancelled()) {
+                            this.grassEaterEntity.eatGrassBonus();
+                            return;
+                        }
+                        //Lava-test end
                         this.entityWorld.playEvent(2001, blockpos1, Block.getIdFromBlock(Blocks.GRASS));
                         this.entityWorld.setBlockState(blockpos1, Blocks.DIRT.getDefaultState(), 2);
                     }
