--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIHurtByTarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIHurtByTarget.java
@@ -28,6 +28,7 @@
 
     public void startExecuting()
     {
+        org.lavapowered.lava.events.LavaEventHandler.targetReason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY; //Lava-test - add TargetReason for event
         this.taskOwner.setAttackTarget(this.taskOwner.getRevengeTarget());
         this.target = this.taskOwner.getAttackTarget();
         this.revengeTimerOld = this.taskOwner.getRevengeTimer();
@@ -70,6 +71,7 @@
 
     protected void setEntityAttackTarget(EntityCreature creatureIn, EntityLivingBase entityLivingBaseIn)
     {
+        org.lavapowered.lava.events.LavaEventHandler.targetReason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY; //Lava-test - add TargetReason for event
         creatureIn.setAttackTarget(entityLivingBaseIn);
     }
 }
