--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIOwnerHurtByTarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIOwnerHurtByTarget.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.passive.EntityTameable;
+import net.minecraft.entity.player.EntityPlayer;
 
 public class EntityAIOwnerHurtByTarget extends EntityAITarget
 {
@@ -41,6 +42,7 @@
 
     public void startExecuting()
     {
+        org.lavapowered.lava.events.LavaEventHandler.targetReason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER; //Lava-test - add TargetReason for event
         this.taskOwner.setAttackTarget(this.attacker);
         EntityLivingBase entitylivingbase = this.tameable.getOwner();
 
