--- ../src-base/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
@@ -38,7 +38,7 @@
     public int fallTime;
     public boolean shouldDropItem = true;
     private boolean dontSetBlock;
-    private boolean hurtEntities;
+    public boolean hurtEntities; //Craftbukkit private -> public
     private int fallHurtMax = 40;
     private float fallHurtAmount = 2.0F;
     public NBTTagCompound tileEntityData;
@@ -114,7 +114,10 @@
             {
                 BlockPos blockpos = new BlockPos(this);
 
-                if (this.world.getBlockState(blockpos).getBlock() == block)
+                //Craftbukkit start - event
+                //if (this.world.getBlockState(blockpos).getBlock() == block)
+                if (this.world.getBlockState(blockpos).getBlock() == block && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR, 0).isCancelled())
+                //Craftbukkit end
                 {
                     this.world.setBlockToAir(blockpos);
                 }
@@ -170,7 +173,7 @@
                     if (!flag1 && BlockFalling.canFallThrough(this.world.getBlockState(new BlockPos(this.posX, this.posY - 0.009999999776482582D, this.posZ))))
                     {
                         this.onGround = false;
-                        return;
+//                        return; // CraftBukkit ///Lava-test why is this here? TODO
                     }
 
                     this.motionX *= 0.699999988079071D;
@@ -185,6 +188,12 @@
                         {
                             if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, this) && (flag1 || !BlockFalling.canFallThrough(this.world.getBlockState(blockpos1.down()))) && this.world.setBlockState(blockpos1, this.fallTile, 3))
                             {
+                                // CraftBukkit start
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos1, this.fallTile.getBlock(), this.fallTile.getBlock().getMetaFromState(this.fallTile)).isCancelled()) {
+                                    return;
+                                }
+                                this.world.setBlockState(blockpos1, this.fallTile, 3);
+                                // CraftBukkit end
                                 if (block instanceof BlockFalling)
                                 {
                                     ((BlockFalling)block).onEndFalling(this.world, blockpos1, this.fallTile, iblockstate);
@@ -248,7 +257,9 @@
 
                 for (Entity entity : list)
                 {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.entityDamage = this; // CraftBukkit
                     entity.attackEntityFrom(damagesource, (float)Math.min(MathHelper.floor((float)i * this.fallHurtAmount), this.fallHurtMax));
+                    org.bukkit.craftbukkit.event.CraftEventFactory.entityDamage = null; // CraftBukkit
                 }
 
                 if (flag && (double)this.rand.nextFloat() < 0.05000000074505806D + (double)i * 0.05D)
