--- ../src-base/minecraft/net/minecraft/entity/item/EntityBoat.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityBoat.java
@@ -64,6 +64,13 @@
     private EntityBoat.Status previousStatus;
     private double lastYd;
 
+    //Craftbukkit start - Some of these haven't worked since a few updates, and since 1.9 they are less and less applicable.
+    public double maxSpeed = 0.4D;
+    public double occupiedDeceleration = 0.2D;
+    public double unoccupiedDeceleration = -1;
+    public boolean landBoats = false;
+    //Craftbukkit end
+
     public EntityBoat(World worldIn)
     {
         super(worldIn);
@@ -138,6 +145,18 @@
             }
             else
             {
+                //Craftbukkit start - event
+                org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+                org.bukkit.entity.Entity attacker = (source.getTrueSource() == null) ? null : source.getTrueSource().getBukkitEntity();
+
+                org.bukkit.event.vehicle.VehicleDamageEvent event = new org.bukkit.event.vehicle.VehicleDamageEvent(vehicle, attacker, (double) amount);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return true;
+                }
+                // amount = event.getDamage(); // TODO Why don't we do this?
+                //Craftbukkit end
                 this.setForwardDirection(-this.getForwardDirection());
                 this.setTimeSinceHit(10);
                 this.setDamageTaken(this.getDamageTaken() + amount * 10.0F);
@@ -146,6 +165,15 @@
 
                 if (flag || this.getDamageTaken() > 40.0F)
                 {
+                    //Craftbukkit start - event
+                    org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, attacker);
+                    this.world.getServer().getPluginManager().callEvent(destroyEvent);
+
+                    if (destroyEvent.isCancelled()) {
+                        this.setDamageTaken(40F); // Maximize damage so this doesn't get triggered again right away
+                        return true;
+                    }
+                    //Craftbukkit end
                     if (!flag && this.world.getGameRules().getBoolean("doEntityDrops"))
                     {
                         this.dropItemWithOffset(this.getItemBoat(), 1, 0.0F);
@@ -169,11 +197,27 @@
         {
             if (entityIn.getEntityBoundingBox().minY < this.getEntityBoundingBox().maxY)
             {
+                //Craftbukkit start - event
+                org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entityIn.getBukkitEntity());
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+                //Craftbukkit end
                 super.applyEntityCollision(entityIn);
             }
         }
         else if (entityIn.getEntityBoundingBox().minY <= this.getEntityBoundingBox().minY)
         {
+            //Craftbukkit start - event
+            org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entityIn.getBukkitEntity());
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
+            //Craftbukkit end
             super.applyEntityCollision(entityIn);
         }
     }
@@ -227,6 +271,7 @@
         return this.getHorizontalFacing().rotateY();
     }
 
+    private org.bukkit.Location lastLocation; //Craftbukkit - add field
     public void onUpdate()
     {
         this.previousStatus = this.status;
@@ -286,6 +331,22 @@
             this.motionZ = 0.0D;
         }
 
+        //Craftbukkit start - event
+        org.bukkit.Server server = this.world.getServer();
+        org.bukkit.World bworld = this.world.getWorld();
+
+        org.bukkit.Location to = new org.bukkit.Location(bworld, this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
+        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+
+        server.getPluginManager().callEvent(new org.bukkit.event.vehicle.VehicleUpdateEvent(vehicle));
+
+        if (lastLocation != null && !lastLocation.equals(to)) {
+            org.bukkit.event.vehicle.VehicleMoveEvent event = new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, lastLocation, to);
+            server.getPluginManager().callEvent(event);
+        }
+        lastLocation = vehicle.getLocation();
+        //Craftbukkit end
+
         for (int i = 0; i <= 1; ++i)
         {
             if (this.getPaddleState(i))
@@ -850,6 +911,15 @@
 
                     if (!this.world.isRemote && !this.isDead)
                     {
+                        //Craftbukkit start - add event ///Lava-test - enhance to preserve original code block
+                        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+                        org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, null);
+                        this.world.getServer().getPluginManager().callEvent(destroyEvent);
+                        if (destroyEvent.isCancelled()) {
+                            this.fallDistance = 0.0F; //TODO check if this is ok
+                            return;
+                        }
+                        //Craftbukkit end
                         this.setDead();
 
                         if (this.world.getGameRules().getBoolean("doEntityDrops"))
