--- ../src-base/minecraft/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
+++ ../src-work/minecraft/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
@@ -26,6 +26,18 @@
                 this.phase.removeAreaEffect();
             }
 
+            //Craftbukkit start - add event
+            org.bukkit.event.entity.EnderDragonChangePhaseEvent event = new org.bukkit.event.entity.EnderDragonChangePhaseEvent(
+                    (org.bukkit.craftbukkit.entity.CraftEnderDragon) this.dragon.getBukkitEntity(),
+                    (this.phase == null) ? null : org.bukkit.craftbukkit.entity.CraftEnderDragon.getBukkitPhase(this.phase.getType()),
+                    org.bukkit.craftbukkit.entity.CraftEnderDragon.getBukkitPhase(phaseIn)
+            );
+            this.dragon.world.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            phaseIn = org.bukkit.craftbukkit.entity.CraftEnderDragon.getMinecraftPhase(event.getNewPhase());
+            //Craftbukkit end
             this.phase = this.getPhase(phaseIn);
 
             if (!this.dragon.world.isRemote)
