--- ../src-base/minecraft/net/minecraft/entity/boss/EntityWither.java
+++ ../src-work/minecraft/net/minecraft/entity/boss/EntityWither.java
@@ -255,15 +255,41 @@
 
             if (j1 <= 0)
             {
-                this.world.newExplosion(this, this.posX, this.posY + (double)this.getEyeHeight(), this.posZ, 7.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this));
-                this.world.playBroadcastSound(1023, new BlockPos(this), 0);
+                //Craftbukkit start - event
+                // this.world.newExplosion(this, this.posX, this.posY + (double)this.getEyeHeight(), this.posZ, 7.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this));
+                // this.world.playBroadcastSound(1023, new BlockPos(this), 0); - changed ???
+                org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 7.0F, false);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    this.world.newExplosion(this, this.posX, this.posY + (double) this.getEyeHeight(), this.posZ, event.getRadius(), event.getFire(), net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this));
+                }
+                
+                int viewDistance = ((net.minecraft.world.WorldServer) this.world).getServer().getViewDistance() * 16;
+                for (EntityPlayerMP player : (List<EntityPlayerMP>) net.minecraft.server.MinecraftServer.getServerCB().getPlayerList().getPlayers()) {
+                    double deltaX = this.posX - player.posX;
+                    double deltaZ = this.posZ - player.posZ;
+                    double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                    if (distanceSquared > viewDistance * viewDistance) {
+                        double deltaLength = Math.sqrt(distanceSquared);
+                        double relativeX = player.posX + (deltaX / deltaLength) * viewDistance;
+                        double relativeZ = player.posZ + (deltaZ / deltaLength) * viewDistance;
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1023, new BlockPos((int) relativeX, (int) this.posY, (int) relativeZ), 0, true));
+                    } else {
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1023, new BlockPos((int) this.posX, (int) this.posY, (int) this.posZ), 0, true));
+                    }
+                }
+                //Craftbukkit end
             }
 
             this.setInvulTime(j1);
 
             if (this.ticksExisted % 10 == 0)
             {
-                this.heal(10.0F);
+                //Craftbukkit start - add RegainReason
+                //this.heal(10.0F);
+                this.heal(10.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.WITHER_SPAWN);
+                //Cratbukkit end
             }
         }
         else
@@ -384,6 +410,11 @@
 
                                 if (!block.isAir(iblockstate, this.world, blockpos) && block.canEntityDestroy(iblockstate, world, blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                                 {
+                                    //Craftbukkit start - add event TODO move to forge?
+                                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR, 0).isCancelled()) {
+                                        continue;
+                                    }
+                                    //Craftbukkit end
                                     flag = this.world.destroyBlock(blockpos, true) || flag;
                                 }
                             }
@@ -399,7 +430,10 @@
 
             if (this.ticksExisted % 20 == 0)
             {
-                this.heal(1.0F);
+                //Craftbukkit start - add RegainReason
+                //this.heal(1.0F);
+                this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN);
+                //Craftbukkit end
             }
 
             this.bossInfo.setPercent(this.getHealth() / this.getMaxHealth());
