--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/EntityRegistry.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/EntityRegistry.java
@@ -147,6 +147,8 @@
     private final ListMultimap<ModContainer, EntityRegistration> entityRegistrations = ArrayListMultimap.create();
     private final BiMap<Class<? extends Entity>, EntityRegistration> entityClassRegistrations = HashBiMap.create();
     private final Map<Class<? extends Entity>, EntityEntry> entityClassEntries = GameData.getEntityClassMap();
+    public static Map<Class<? extends Entity>, String> entityTypeMap = com.google.common.collect.Maps.newHashMap(); //Lava Used by CraftCustomEntity
+    public static Map<String, Class<? extends Entity>> entityClassMap = com.google.common.collect.Maps.newHashMap(); //Lava Use by CraftWorld
 
     public static EntityRegistry instance()
     {
@@ -171,6 +173,7 @@
     public static void registerModEntity(ResourceLocation registryName, Class<? extends Entity> entityClass, String entityName, int id, Object mod, int trackingRange, int updateFrequency, boolean sendsVelocityUpdates)
     {
         instance().doModEntityRegistration(registryName, entityClass, entityName, id, mod, trackingRange, updateFrequency, sendsVelocityUpdates);
+        registerBukkitType(entityClass, entityName); //Lava - register the entity in bukkit too
     }
 
     /**
@@ -191,6 +194,7 @@
     {
         instance().doModEntityRegistration(registryName, entityClass, entityName, id, mod, trackingRange, updateFrequency, sendsVelocityUpdates);
         EntityRegistry.registerEgg(registryName, eggPrimary, eggSecondary);
+        registerBukkitType(entityClass, entityName); //Lava - register in bukkit too
     }
 
     private void doModEntityRegistration(ResourceLocation registryName, Class<? extends Entity> entityClass, String entityName, int id, Object mod, int trackingRange, int updateFrequency, boolean sendsVelocityUpdates)
@@ -385,4 +389,28 @@
         this.entityClassRegistrations.put(entity, registration);
         this.entityRegistrations.put(registration.container, registration);
     }
+
+    //Lava-test method - register entities to bukkit as well TODO check this
+    private static void registerBukkitType(Class<? extends Entity> entityClass, String entityName) {
+        /*ModContainer activeModContainer = net.minecraftforge.fml.common.Loader.instance().activeModContainer();
+        String modId = "unknown";
+        // Fixup bad entity names from mods TODO - nope, we want the entity name provided by mod
+        if (entityName.contains(".")) {
+            if ((entityName.indexOf("." + 1) < entityName.length())) {
+                entityName = entityName.substring(entityName.indexOf(".") + 1);
+            }
+        }
+        entityName = entityName.replace("entity", "");
+        if (entityName.startsWith("ent")) {
+            entityName = entityName.replace("ent", "");
+        }
+        entityName = entityName.replaceAll("[^A-Za-z0-9]", ""); // Remove all non-digits/alphanumeric
+        if (activeModContainer != null) {
+            modId = activeModContainer.getModId();
+            entityName = modId + "-" + entityName;*/
+        entityName = entityName.replace(" ", ""); //replace spaces for commands
+        entityTypeMap.put(entityClass, entityName);
+        entityClassMap.put(entityName, entityClass);
+        //}
+    }
 }
