--- ../src-base/minecraft/net/minecraft/entity/monster/EntityVex.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityVex.java
@@ -316,6 +316,7 @@
 
         public void startExecuting()
         {
+            org.lavapowered.lava.events.LavaEventHandler.targetReason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET; //Lava-test - add TargetReason
             EntityVex.this.setAttackTarget(EntityVex.this.owner.getAttackTarget());
             super.startExecuting();
         }
