--- ../src-base/minecraft/net/minecraft/entity/monster/EntityVex.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityVex.java
@@ -316,7 +316,10 @@
 
         public void startExecuting()
         {
-            EntityVex.this.setAttackTarget(EntityVex.this.owner.getAttackTarget());
+            //Craftbukkit start - add target reason
+            // EntityVex.this.setAttackTarget(EntityVex.this.owner.getAttackTarget());
+            EntityVex.this.setAttackTarget(EntityVex.this.owner.getAttackTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true);
+            //Craftbukkit end
             super.startExecuting();
         }
     }
