--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
@@ -221,6 +221,11 @@
 
                     if (BlockSilverfish.canContainSilverfish(iblockstate))
                     {
+                        //Craftbukkit start - event
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.entity, blockpos, Blocks.MONSTER_EGG, Block.getIdFromBlock(BlockSilverfish.getBlockById(iblockstate.getBlock().getMetaFromState(iblockstate)))).isCancelled()) {
+                            return;
+                        }
+                        //Craftbukkit end
                         world.setBlockState(blockpos, Blocks.MONSTER_EGG.getDefaultState().withProperty(BlockSilverfish.VARIANT, BlockSilverfish.EnumType.forModelBlock(iblockstate)), 3);
                         this.entity.spawnExplosionParticle();
                         this.entity.setDead();
@@ -273,6 +278,11 @@
 
                                 if (iblockstate.getBlock() == Blocks.MONSTER_EGG)
                                 {
+                                    //Craftbukkit start - event
+                                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockpos1, Blocks.AIR, 0).isCancelled()) {
+                                        continue;
+                                    }
+                                    //Craftbukkit end
                                     if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this.silverfish))
                                     {
                                         world.destroyBlock(blockpos1, true);
