--- ../src-base/minecraft/net/minecraft/entity/monster/EntityZombie.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityZombie.java
@@ -215,7 +215,15 @@
 
                 if (flag)
                 {
-                    this.setFire(8);
+                    //Craftbukkit start - event
+                    // this.setFire(8);
+                    org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        this.setFire(event.getDuration());
+                    }
+                    //Craftbukkit end
                 }
             }
         }
@@ -270,8 +278,13 @@
 
                         if (!this.world.isAnyPlayerWithinRangeAt((double)i1, (double)j1, (double)k1, 7.0D) && this.world.checkNoEntityCollision(entityzombie.getEntityBoundingBox(), entityzombie) && this.world.getCollisionBoxes(entityzombie, entityzombie.getEntityBoundingBox()).isEmpty() && !this.world.containsAnyLiquid(entityzombie.getEntityBoundingBox()))
                         {
-                            this.world.spawnEntity(entityzombie);
+                            //Craftbukkit start - add reason
+                            //this.world.spawnEntity(entityzombie);
+                            this.world.spawnEntity(entityzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.REINFORCEMENTS);
+                            //Craftbukkit end
+                            org.lavapowered.lava.events.LavaEventHandler.targetReason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.REINFORCEMENT_TARGET; //Lava-test - add TargetReason
                             if (entitylivingbase != null) entityzombie.setAttackTarget(entitylivingbase);
+                            org.lavapowered.lava.events.LavaEventHandler.targetReason = null; //Lava-test - reset as it's not guaranteed that we will call the event
                             entityzombie.onInitialSpawn(this.world.getDifficultyForLocation(new BlockPos(entityzombie)), (IEntityLivingData)null);
                             this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement caller charge", -0.05000000074505806D, 0));
                             entityzombie.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
@@ -299,7 +312,15 @@
 
             if (this.getHeldItemMainhand().isEmpty() && this.isBurning() && this.rand.nextFloat() < f * 0.3F)
             {
-                entityIn.setFire(2 * (int)f);
+                //Craftbukkit start - event
+                // entityIn.setFire(2 * (int)f);
+                org.bukkit.event.entity.EntityCombustByEntityEvent event = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entityIn.getBukkitEntity(), 2 * (int) f); // PAIL: fixme
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    entityIn.setFire(event.getDuration());
+                }
+                //Craftbukkit end
             }
         }
 
@@ -416,7 +437,10 @@
                 entityzombievillager.setAlwaysRenderNameTag(entityvillager.getAlwaysRenderNameTag());
             }
 
-            this.world.spawnEntity(entityzombievillager);
+            //Craftbukkit start - add spawn reason
+            //this.world.spawnEntity(entityzombievillager);
+            this.world.spawnEntity(entityzombievillager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.INFECTION);
+            //Craftbukkit end
             this.world.playEvent((EntityPlayer)null, 1026, new BlockPos(this), 0);
         }
     }
@@ -443,7 +467,7 @@
     {
         livingdata = super.onInitialSpawn(difficulty, livingdata);
         float f = difficulty.getClampedAdditionalDifficulty();
-        this.setCanPickUpLoot(this.rand.nextFloat() < 0.55F * f);
+        this.idkwhyreyoudoingthis(this.rand.nextFloat() < 0.55F * f);
 
         if (livingdata == null)
         {
@@ -475,7 +499,10 @@
                     entitychicken1.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, 0.0F);
                     entitychicken1.onInitialSpawn(difficulty, (IEntityLivingData)null);
                     entitychicken1.setChickenJockey(true);
-                    this.world.spawnEntity(entitychicken1);
+                    //Craftbukkit start - add spawn reason
+                    //this.world.spawnEntity(entitychicken1);
+                    this.world.spawnEntity(entitychicken1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.MOUNT);
+                    //Craftbukkit end
                     this.startRiding(entitychicken1);
                 }
             }
@@ -543,7 +570,7 @@
 
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+        // super.onDeath(cause); // CraftBukkit - moved down
 
         if (cause.getTrueSource() instanceof EntityCreeper)
         {
@@ -560,6 +587,7 @@
                 }
             }
         }
+        super.onDeath(cause); //Craftbukkit - moved up
     }
 
     protected ItemStack getSkullDrop()
