--- ../src-base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCactus.java
@@ -58,6 +58,8 @@
                 {
                     worldIn.setBlockState(blockpos, this.getDefaultState());
                     IBlockState iblockstate = state.withProperty(AGE, Integer.valueOf(0));
+                    //Lava-test - move bukkit event to ForgeHook
+                    //org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(worldIn, blockpos.getX(), blockpos.getY(), blockpos.getZ(), this, 0); //Craftbukkit - event
                     worldIn.setBlockState(pos, iblockstate, 4);
                     iblockstate.neighborChanged(worldIn, blockpos, this, pos);
                 }
@@ -123,7 +125,10 @@
 
     public void onEntityCollidedWithBlock(World worldIn, BlockPos pos, IBlockState state, Entity entityIn)
     {
+    	//Lava-test - try to not use this field TODO 
+        //org.bukkit.craftbukkit.event.CraftEventFactory.blockDamage = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()); //Craftbukkit - used in EntityDamageByBlockEvent event
         entityIn.attackEntityFrom(DamageSource.CACTUS, 1.0F);
+        //org.bukkit.craftbukkit.event.CraftEventFactory.blockDamage = null; //Craftbukkit - used in EntityDamageByBlockEvent event
     }
 
     public IBlockState getStateFromMeta(int meta)
