--- ../src-base/minecraft/net/minecraft/block/BlockGrass.java
+++ ../src-work/minecraft/net/minecraft/block/BlockGrass.java
@@ -41,7 +41,18 @@
             if (!worldIn.isAreaLoaded(pos, 3)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
             if (worldIn.getLightFromNeighbors(pos.up()) < 4 && worldIn.getBlockState(pos.up()).getLightOpacity(worldIn, pos.up()) > 2)
             {
-                worldIn.setBlockState(pos, Blocks.DIRT.getDefaultState());
+            	//Craftbukkit start - call event
+//                worldIn.setBlockState(pos, Blocks.DIRT.getDefaultState());
+                org.bukkit.block.BlockState blockState = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()).getState();
+                blockState.setType(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(Blocks.DIRT));
+
+                org.bukkit.event.block.BlockFadeEvent event = new org.bukkit.event.block.BlockFadeEvent(blockState.getBlock(), blockState);
+                worldIn.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    blockState.update(true);
+                }
+                //Craftbukkit end
             }
             else
             {
@@ -61,7 +72,19 @@
 
                         if (iblockstate1.getBlock() == Blocks.DIRT && iblockstate1.getValue(BlockDirt.VARIANT) == BlockDirt.DirtType.DIRT && worldIn.getLightFromNeighbors(blockpos.up()) >= 4 && iblockstate.getLightOpacity(worldIn, pos.up()) <= 2)
                         {
-                            worldIn.setBlockState(blockpos, Blocks.GRASS.getDefaultState());
+                        	//Craftbukkit start - call event
+//                            worldIn.setBlockState(blockpos, Blocks.GRASS.getDefaultState());
+                            org.bukkit.World bworld = worldIn.getWorld();
+                            org.bukkit.block.BlockState blockState = bworld.getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()).getState();
+                            blockState.setType(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(Blocks.GRASS));
+
+                            org.bukkit.event.block.BlockSpreadEvent event = new org.bukkit.event.block.BlockSpreadEvent(blockState.getBlock(), bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()), blockState);
+                            worldIn.getServer().getPluginManager().callEvent(event);
+
+                            if (!event.isCancelled()) {
+                                blockState.update(true);
+                            }
+                            //Craftbukkit end
                         }
                     }
                 }
@@ -109,7 +132,10 @@
 
                             if (Blocks.TALLGRASS.canBlockStay(worldIn, blockpos1, iblockstate1))
                             {
-                                worldIn.setBlockState(blockpos1, iblockstate1, 3);
+                            	//Craftbukkit start - call event
+//                                worldIn.setBlockState(blockpos1, iblockstate1, 3);
+                                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(worldIn, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ(), iblockstate1.getBlock(), iblockstate1.getBlock().getMetaFromState(iblockstate1));
+                                //Craftbukkit end
                             }
                         }
                     }
