--- ../src-base/minecraft/net/minecraft/block/BlockLiquid.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLiquid.java
@@ -286,6 +286,10 @@
                 {
                     worldIn.setBlockState(pos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, pos, pos, Blocks.OBSIDIAN.getDefaultState()));
                     this.triggerMixEffects(worldIn, pos);
+                    //Lava-test move this event to ForgeEventFactory
+                    /*if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(worldIn, pos, Blocks.OBSIDIAN.getDefaultState(), null)) {
+                        this.triggerMixEffects(worldIn, pos);
+                    }*/
                     return true;
                 }
 
@@ -293,6 +297,10 @@
                 {
                     worldIn.setBlockState(pos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, pos, pos, Blocks.COBBLESTONE.getDefaultState()));
                     this.triggerMixEffects(worldIn, pos);
+                    //Lava-test move this event to ForgeEventFactory
+                    /*if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(worldIn, pos, Blocks.COBBLESTONE.getDefaultState(), null)) {
+                        this.triggerMixEffects(worldIn, pos);
+                    }*/
                     return true;
                 }
             }
