--- ../src-base/minecraft/net/minecraft/block/BlockReed.java
+++ ../src-work/minecraft/net/minecraft/block/BlockReed.java
@@ -60,6 +60,9 @@
                     if (j == 15)
                     {
                         worldIn.setBlockState(pos.up(), this.getDefaultState());
+                        //Lava-test - move to ForgeHooks
+                        //BlockPos upPos = pos.up();
+                        //org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(worldIn, upPos.getX(), upPos.getY(), upPos.getZ(), this, 0);
                         worldIn.setBlockState(pos, state.withProperty(AGE, Integer.valueOf(0)), 4);
                     }
                     else
