--- ../src-base/minecraft/net/minecraft/block/BlockDragonEgg.java
+++ ../src-work/minecraft/net/minecraft/block/BlockDragonEgg.java
@@ -97,6 +97,16 @@
 
                 if (worldIn.isAirBlock(blockpos))
                 {
+                	//Craftbukkit start - event
+                    org.bukkit.event.block.BlockFromToEvent event = new org.bukkit.event.block.BlockFromToEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), worldIn.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                    org.bukkit.Bukkit.getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return;
+                    }
+
+                    blockpos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+                    //Craftbukkit end
                     if (worldIn.isRemote)
                     {
                         for (int j = 0; j < 128; ++j)
