--- ../src-base/minecraft/net/minecraft/block/BlockCommandBlock.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCommandBlock.java
@@ -58,6 +58,14 @@
                 TileEntityCommandBlock tileentitycommandblock = (TileEntityCommandBlock)tileentity;
                 boolean flag = worldIn.isBlockPowered(pos);
                 boolean flag1 = tileentitycommandblock.isPowered();
+                //Craftbukkit start - redstone event
+                int old = flag1 ? 15 : 0;
+                int current = flag ? 15 : 0;
+
+                org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), old, current);
+                worldIn.getServer().getPluginManager().callEvent(eventRedstone);
+                flag = eventRedstone.getNewCurrent() > 0;
+                //Craftbukkit end
                 tileentitycommandblock.setPowered(flag);
 
                 if (!flag1 && !tileentitycommandblock.isAuto() && tileentitycommandblock.getMode() != TileEntityCommandBlock.Mode.SEQUENCE)
