--- ../src-base/minecraft/net/minecraft/block/BlockStaticLiquid.java
+++ ../src-work/minecraft/net/minecraft/block/BlockStaticLiquid.java
@@ -63,6 +63,14 @@
                         {
                             if (this.isSurroundingBlockFlammable(worldIn, blockpos))
                             {
+                            	//Craftbukkit start - event
+                            	//Lava-test - move to ForgeEventFactory
+                                /*if (worldIn.getBlockState(blockpos) != Blocks.FIRE) {
+                                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(worldIn, blockpos.getX(), blockpos.getY(), blockpos.getZ(), pos.getX(), pos.getY(), pos.getZ()).isCancelled()) {
+                                        continue;
+                                    }
+                                }*/
+                                //Craftbukkit end
                                 worldIn.setBlockState(blockpos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, blockpos, pos, Blocks.FIRE.getDefaultState()));
                                 return;
                             }
@@ -86,6 +94,15 @@
 
                         if (worldIn.isAirBlock(blockpos1.up()) && this.getCanBlockBurn(worldIn, blockpos1))
                         {
+                            // CraftBukkit start - Prevent lava putting something on fire
+                        	//Lava-test - move to ForgeEventFactory
+                            /*BlockPos up = blockpos1.up();
+                            if (worldIn.getBlockState(up) != Blocks.FIRE) {
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(worldIn, up.getX(), up.getY(), up.getZ(), pos.getX(), pos.getY(), pos.getZ()).isCancelled()) {
+                                    continue;
+                                }
+                            }*/
+                            // CraftBukkit end
                             worldIn.setBlockState(blockpos1.up(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, blockpos1.up(), pos, Blocks.FIRE.getDefaultState()));
                         }
                     }
