--- ../src-base/minecraft/net/minecraft/block/BlockSkull.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSkull.java
@@ -128,6 +128,31 @@
         return new ItemStack(Items.SKULL, 1, i);
     }
 
+    // CraftBukkit start - Special case dropping so we can get info from the tile entity
+    //Lava-test - move everything back, I have no clue why is this here TODO investigate
+    /*@Override
+    public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune) {
+        if (worldIn.rand.nextFloat() < chance) {
+            TileEntity tileentity = worldIn.getTileEntity(pos);
+
+            if (tileentity instanceof TileEntitySkull) {
+                TileEntitySkull tileentityskull = (TileEntitySkull) tileentity;
+                ItemStack itemstack = this.getItem(worldIn, pos, state);
+
+                if (tileentityskull.getSkullType() == 3 && tileentityskull.getPlayerProfile() != null) {
+                    itemstack.setTagCompound(new NBTTagCompound());
+                    NBTTagCompound nbttagcompound = new NBTTagCompound();
+
+                    NBTUtil.writeGameProfile(nbttagcompound, tileentityskull.getPlayerProfile());
+                    itemstack.getTagCompound().setTag("SkullOwner", nbttagcompound);
+                }
+
+                spawnAsEntity(worldIn, pos, itemstack);
+            }
+        }
+    }*/
+    // CraftBukkit end
+
     public void onBlockHarvested(World worldIn, BlockPos pos, IBlockState state, EntityPlayer player)
     {
         if (player.capabilities.isCreativeMode)
@@ -148,6 +173,7 @@
     {
         {
             if (!((Boolean)state.getValue(NODROP)).booleanValue())
+            //if (false) CraftBukkit - Drop item in code above, not here /// Lava-test - revert
             {
                 TileEntity tileentity = worldIn.getTileEntity(pos);
 
@@ -189,6 +215,7 @@
 
     public void checkWitherSpawn(World worldIn, BlockPos pos, TileEntitySkull te)
     {
+        if (worldIn.captureBlockStates) return; //Craftbukkit ??? TODO investigate
         if (te.getSkullType() == 1 && pos.getY() >= 2 && worldIn.getDifficulty() != EnumDifficulty.PEACEFUL && !worldIn.isRemote)
         {
             BlockPattern blockpattern = this.getWitherPattern();
@@ -196,10 +223,16 @@
 
             if (blockpattern$patternhelper != null)
             {
+                org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(worldIn.getWorld()); //Craftbukkit - prepare for event?
                 for (int i = 0; i < 3; ++i)
                 {
                     BlockWorldState blockworldstate = blockpattern$patternhelper.translateOffset(i, 0, 0);
-                    worldIn.setBlockState(blockworldstate.getPos(), blockworldstate.getBlockState().withProperty(NODROP, Boolean.valueOf(true)), 2);
+                    //Craftbukkit start - prepare for event?
+                    BlockPos pos1 = blockworldstate.getPos();
+                    IBlockState data = blockworldstate.getBlockState().withProperty(BlockSkull.NODROP, true);
+                    blockList.setTypeAndData(pos1.getX(), pos1.getY(), pos1.getZ(), data.getBlock(), data.getBlock().getMetaFromState(data), 2);
+//                    worldIn.setBlockState(blockworldstate.getPos(), blockworldstate.getBlockState().withProperty(NODROP, Boolean.valueOf(true)), 2);
+                    //Craftbukkit end
                 }
 
                 for (int j = 0; j < blockpattern.getPalmLength(); ++j)
@@ -207,7 +240,11 @@
                     for (int k = 0; k < blockpattern.getThumbLength(); ++k)
                     {
                         BlockWorldState blockworldstate1 = blockpattern$patternhelper.translateOffset(j, k, 0);
-                        worldIn.setBlockState(blockworldstate1.getPos(), Blocks.AIR.getDefaultState(), 2);
+                        //Craftbukkit start - prepare for event?
+                        BlockPos pos1 = blockworldstate1.getPos();
+                        blockList.setTypeAndData(pos1.getX(), pos1.getY(), pos1.getZ(), Blocks.AIR, 0, 2);
+//                        worldIn.setBlockState(blockworldstate1.getPos(), Blocks.AIR.getDefaultState(), 2);
+                        //Craftbukkit end
                     }
                 }
 
@@ -217,13 +254,19 @@
                 entitywither.setLocationAndAngles((double)blockpos1.getX() + 0.5D, (double)blockpos1.getY() + 0.55D, (double)blockpos1.getZ() + 0.5D, blockpattern$patternhelper.getForwards().getAxis() == EnumFacing.Axis.X ? 0.0F : 90.0F, 0.0F);
                 entitywither.renderYawOffset = blockpattern$patternhelper.getForwards().getAxis() == EnumFacing.Axis.X ? 0.0F : 90.0F;
                 entitywither.ignite();
-
+                //Craftbukkit start - add condition for event? /// Lava-test - invert condition to restore original code block
+                if (!worldIn.spawnEntity(entitywither, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
+                	return;
+                }
+                
+                blockList.updateList();
+                //Craftbukkit end
                 for (EntityPlayerMP entityplayermp : worldIn.getEntitiesWithinAABB(EntityPlayerMP.class, entitywither.getEntityBoundingBox().grow(50.0D)))
                 {
                     CriteriaTriggers.SUMMONED_ENTITY.trigger(entityplayermp, entitywither);
                 }
 
-                worldIn.spawnEntity(entitywither);
+//                worldIn.spawnEntity(entitywither);
 
                 for (int l = 0; l < 120; ++l)
                 {
