--- ../src-base/minecraft/net/minecraft/block/BlockTripWireHook.java
+++ ../src-work/minecraft/net/minecraft/block/BlockTripWireHook.java
@@ -196,7 +196,16 @@
             this.notifyNeighbors(worldIn, blockpos1, enumfacing1);
             this.playSound(worldIn, blockpos1, flag2, flag3, flag, flag1);
         }
+        
+        //Craftbukkit start - event
+        org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), 15, 0);
+        worldIn.getServer().getPluginManager().callEvent(eventRedstone);
 
+        if (eventRedstone.getNewCurrent() > 0) {
+            return;
+        }
+        //Craftbukkit end
+
         this.playSound(worldIn, pos, flag2, flag3, flag, flag1);
 
         if (!p_176260_4_)
