--- ../src-base/minecraft/net/minecraft/block/BlockRedstoneLight.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedstoneLight.java
@@ -30,10 +30,20 @@
         {
             if (this.isOn && !worldIn.isBlockPowered(pos))
             {
+            	//Craftbukkit start - event
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(worldIn, pos.getX(), pos.getY(), pos.getZ(), 15, 0).getNewCurrent() != 0) {
+                    return;
+                }
+                //Craftbukkit end
                 worldIn.setBlockState(pos, Blocks.REDSTONE_LAMP.getDefaultState(), 2);
             }
             else if (!this.isOn && worldIn.isBlockPowered(pos))
             {
+            	//Craftbukkit start - event
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(worldIn, pos.getX(), pos.getY(), pos.getZ(), 0, 15).getNewCurrent() != 15) {
+                    return;
+                }
+                //Craftbukkit end
                 worldIn.setBlockState(pos, Blocks.LIT_REDSTONE_LAMP.getDefaultState(), 2);
             }
         }
@@ -49,6 +59,11 @@
             }
             else if (!this.isOn && worldIn.isBlockPowered(pos))
             {
+            	//Craftbukkit start - event
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(worldIn, pos.getX(), pos.getY(), pos.getZ(), 0, 15).getNewCurrent() != 15) {
+                    return;
+                }
+                //Craftbukkit end
                 worldIn.setBlockState(pos, Blocks.LIT_REDSTONE_LAMP.getDefaultState(), 2);
             }
         }
@@ -60,6 +75,11 @@
         {
             if (this.isOn && !worldIn.isBlockPowered(pos))
             {
+            	//Craftbukkit start - event
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(worldIn, pos.getX(), pos.getY(), pos.getZ(), 15, 0).getNewCurrent() != 0) {
+                    return;
+                }
+                //Craftbukkit end
                 worldIn.setBlockState(pos, Blocks.REDSTONE_LAMP.getDefaultState(), 2);
             }
         }
