--- ../src-base/minecraft/net/minecraft/block/BlockPortal.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPortal.java
@@ -74,8 +74,11 @@
 
             if (i > 0 && !worldIn.getBlockState(blockpos.up()).isNormalCube())
             {
-                Entity entity = ItemMonsterPlacer.spawnCreature(worldIn, EntityList.getKey(EntityPigZombie.class), (double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 1.1D, (double)blockpos.getZ() + 0.5D);
-
+            	//Craftbukkit start - add spawnreason
+            	//Entity entity = ItemMonsterPlacer.spawnCreature(worldIn, EntityList.getKey(EntityPigZombie.class), (double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 1.1D, (double)blockpos.getZ() + 0.5D);
+                Entity entity = ItemMonsterPlacer.spawnCreature(worldIn, EntityList.getKey(EntityPigZombie.class), (double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 1.1D, (double)blockpos.getZ() + 0.5D, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NETHER_PORTAL);
+                //Craftbukkit end
+                
                 if (entity != null)
                 {
                     entity.timeUntilPortal = entity.getPortalCooldown();
@@ -216,6 +219,10 @@
     {
         if (!entityIn.isRiding() && !entityIn.isBeingRidden() && entityIn.isNonBoss())
         {
+        	//Craftbukkit start - call event
+            org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entityIn.getBukkitEntity(), new org.bukkit.Location(worldIn.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            worldIn.getServer().getPluginManager().callEvent(event);
+            //Craftbukkit end
             entityIn.setPortal(pos);
         }
     }
@@ -370,6 +377,7 @@
             private BlockPos bottomLeft;
             private int height;
             private int width;
+            java.util.Collection<org.bukkit.block.Block> blocks = new java.util.HashSet<org.bukkit.block.Block>(); //Craftbukkit - add field for portal blocks
 
             public Size(World worldIn, BlockPos p_i45694_2_, EnumFacing.Axis p_i45694_3_)
             {
@@ -442,6 +450,10 @@
 
             protected int calculatePortalHeight()
             {
+            	//Craftbukkit start - clear portal blocks
+                this.blocks.clear();
+                org.bukkit.World bworld = this.world.getWorld();
+                //Craftbukkit end
                 label56:
 
                 for (this.height = 0; this.height < 21; ++this.height)
@@ -468,7 +480,12 @@
                             if (block != Blocks.OBSIDIAN)
                             {
                                 break label56;
+                            //Craftbukkit start - add portal blocks
+                            } else {
+                                BlockPos pos = blockpos.offset(this.leftDir);
+                                blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                             }
+                            //Craftbukkit end
                         }
                         else if (i == this.width - 1)
                         {
@@ -477,7 +494,12 @@
                             if (block != Blocks.OBSIDIAN)
                             {
                                 break label56;
+                            //Craftbukkit start - add portal blocks
+                            } else {
+                                BlockPos pos = blockpos.offset(this.rightDir);
+                                blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                             }
+                            //Craftbukkit end
                         }
                     }
                 }
@@ -488,7 +510,12 @@
                     {
                         this.height = 0;
                         break;
+                    //Craftbukkit start - add portal blocks
+                    } else {
+                        BlockPos pos = this.bottomLeft.offset(this.rightDir, j).up(this.height);
+                        blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                     }
+                    //Craftbukkit end
                 }
 
                 if (this.height <= 21 && this.height >= 3)
@@ -526,5 +553,42 @@
                     }
                 }
             }
+            
+            //Craftbukkit start - call event method
+            public boolean placePortalBlocksCB()
+            {
+                org.bukkit.World bworld = this.world.getWorld();
+
+                // Copy below for loop
+                for (int i = 0; i < this.width; ++i) {
+                    BlockPos blockposition = this.bottomLeft.offset(this.rightDir, i);
+
+                    for (int j = 0; j < this.height; ++j) {
+                        BlockPos pos = blockposition.up(j);
+                        blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                    }
+                }
+
+                org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent(blocks, bworld, org.bukkit.event.world.PortalCreateEvent.CreateReason.FIRE);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return false;
+                }
+                
+                //Lava-test - handled later
+                /*for (int i = 0; i < this.width; ++i)
+                {
+                    BlockPos blockpos = this.bottomLeft.offset(this.rightDir, i);
+
+                    for (int j = 0; j < this.height; ++j)
+                    {
+                        this.world.setBlockState(blockpos.up(j), Blocks.PORTAL.getDefaultState().withProperty(BlockPortal.AXIS, this.axis), 2);
+                    }
+                }*/
+
+                return true;
+            }
+            //Craftbukkit end
         }
 }
