--- ../src-base/minecraft/net/minecraft/block/BlockCake.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCake.java
@@ -71,7 +71,19 @@
         else
         {
             player.addStat(StatList.CAKE_SLICES_EATEN);
+            // Lava Restore CraftBukkit changes for AppleCore recognition TODO deeper invetigation
             player.getFoodStats().addStats(2, 0.1F);
+            //Craftbukkit start - event
+            int oldFoodLevel = player.getFoodStats().foodLevel;
+
+            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(player, 2 + oldFoodLevel);
+
+            if (!event.isCancelled()) {
+                player.getFoodStats().addStats(event.getFoodLevel() - oldFoodLevel, 0.1F);
+            }
+
+            ((net.minecraft.entity.player.EntityPlayerMP) player).getBukkitEntity().sendHealthUpdate();
+            //Craftbukkit end
             int i = ((Integer)state.getValue(BITES)).intValue();
 
             if (i < 6)
