--- ../src-base/minecraft/net/minecraft/block/BlockPumpkin.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPumpkin.java
@@ -62,6 +62,7 @@
     private void trySpawnGolem(World worldIn, BlockPos pos)
     {
         BlockPattern.PatternHelper blockpattern$patternhelper = this.getSnowmanPattern().match(worldIn, pos);
+        org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(worldIn.getWorld()); //Craftbukkit
 
         if (blockpattern$patternhelper != null)
         {
@@ -102,7 +103,11 @@
                 {
                     for (int k = 0; k < this.getGolemPattern().getThumbLength(); ++k)
                     {
-                        worldIn.setBlockState(blockpattern$patternhelper.translateOffset(j, k, 0).getPos(), Blocks.AIR.getDefaultState(), 2);
+                    	//Craftbukkit start - event preparation?
+//                        worldIn.setBlockState(blockpattern$patternhelper.translateOffset(j, k, 0).getPos(), Blocks.AIR.getDefaultState(), 2);
+                        BlockPos pos2 = blockpattern$patternhelper.translateOffset(j, k, 0).getPos();
+                        blockList.setTypeId(pos2.getX(), pos2.getY(), pos2.getZ(), 0);
+                        //Craftbukkit end
                     }
                 }
 
@@ -110,7 +115,13 @@
                 EntityIronGolem entityirongolem = new EntityIronGolem(worldIn);
                 entityirongolem.setPlayerCreated(true);
                 entityirongolem.setLocationAndAngles((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.05D, (double)blockpos.getZ() + 0.5D, 0.0F, 0.0F);
-                worldIn.spawnEntity(entityirongolem);
+                //Craftbukkit start - event? /// Lava-test - invert condition to restore original code block 
+                //worldIn.spawnEntity(entityirongolem);
+                if (!worldIn.spawnEntity(entityirongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM)) {
+                	return;
+                }
+                
+                blockList.updateList();
 
                 for (EntityPlayerMP entityplayermp1 : worldIn.getEntitiesWithinAABB(EntityPlayerMP.class, entityirongolem.getEntityBoundingBox().grow(5.0D)))
                 {
