--- ../src-base/minecraft/net/minecraft/block/BlockLeaves.java
+++ ../src-work/minecraft/net/minecraft/block/BlockLeaves.java
@@ -178,6 +178,14 @@
 
     private void destroy(World worldIn, BlockPos pos)
     {
+    	//Craftbukkit start - event
+        org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+        worldIn.getServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled() || worldIn.getBlockState(pos).getBlock() != this) {
+            return;
+        }
+        //Craftbukkit end
         this.dropBlockAsItem(worldIn, pos, worldIn.getBlockState(pos), 0);
         worldIn.setBlockToAir(pos);
     }
