--- ../src-base/minecraft/net/minecraftforge/oredict/ShapedOreRecipe.java
+++ ../src-work/minecraft/net/minecraftforge/oredict/ShapedOreRecipe.java
@@ -254,4 +254,13 @@
         ItemStack result = CraftingHelper.getItemStack(JsonUtils.getJsonObject(json, "result"), context);
         return new ShapedOreRecipe(group.isEmpty() ? null : new ResourceLocation(group), result, primer);
     }
+
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        throw new UnsupportedOperationException("Cannot get Bukkit Recipe from ShapedOreRecipe");
+    }
+
+    @Override
+    public void setKey(ResourceLocation key) {
+    }
 }
