--- ../src-base/minecraft/net/minecraft/item/crafting/CraftingManager.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/CraftingManager.java
@@ -42,6 +42,7 @@
     {
         try
         {
+            CraftingManager.nextAvailableId = 0; //Craftbukkit - Reset recipe ID count TODO ???
             register("armordye", new RecipesArmorDyes());
             register("bookcloning", new RecipeBookCloning());
             register("mapcloning", new RecipesMapCloning());
@@ -188,6 +189,7 @@
         }
         else
         {
+            recipe.setKey(name); //Craftbukkit ???
             REGISTRY.register(nextAvailableId++, name, recipe);
         }
     }
@@ -202,6 +204,7 @@
             }
         }
 
+        craftMatrix.currentRecipe = null; // CraftBukkit - Clear recipe when no recipe is found
         return ItemStack.EMPTY;
     }
 
@@ -212,10 +215,12 @@
         {
             if (irecipe.matches(craftMatrix, worldIn))
             {
+                craftMatrix.currentRecipe = irecipe; // CraftBukkit
                 return irecipe;
             }
         }
 
+        craftMatrix.currentRecipe = null; // CraftBukkit - Clear recipe when no recipe is found
         return null;
     }
 
