--- ../src-base/minecraft/net/minecraft/item/crafting/IRecipe.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/IRecipe.java
@@ -34,4 +34,6 @@
     {
         return "";
     }
+default org.bukkit.inventory.Recipe toBukkitRecipe() { return null; } //Craftbukkit - converting stuff TODO how the mods react to this?
+default void setKey(net.minecraft.util.ResourceLocation key) {} //Craftbukkit
 }
