--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -79,6 +79,18 @@
                     i1 = 0;
                 }
 
+                // CraftBukkit start - Construct a dummy repair recipe
+                ItemStack result = new ItemStack(itemstack3.getItem(), 1, i1);
+                NonNullList<Ingredient> ingredients = NonNullList.create();
+                ingredients.add(Ingredient.fromStacks(new ItemStack[]{itemstack2.copy()}));
+                ingredients.add(Ingredient.fromStacks(new ItemStack[]{itemstack3.copy()}));
+                ShapelessRecipes recipe = new ShapelessRecipes("", result.copy(), ingredients);
+                recipe.key = new net.minecraft.util.ResourceLocation("repairitem");
+                inv.currentRecipe = recipe;
+                result = org.bukkit.craftbukkit.event.CraftEventFactory.callPreCraftEvent(inv, result, inv.eventHandler.getBukkitView(), true);
+                itemstack2 = result;
+                i1 = result.getMetadata();
+                // CraftBukkit end TODO support more than 1 item in output?
                 return new ItemStack(itemstack2.getItem(), 1, i1);
             }
         }
