--- ../src-base/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
@@ -27,6 +27,7 @@
     public final NonNullList<Ingredient> recipeItems;
     private final ItemStack recipeOutput;
     private final String group;
+    public ResourceLocation key; //Craftbukkit TODO find usage for this
 
     public ShapedRecipes(String group, int width, int height, NonNullList<Ingredient> ingredients, ItemStack result)
     {
@@ -373,6 +374,69 @@
         }
     }
 
+    //Craftbukkit start - methods
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.recipeOutput);
+        org.bukkit.craftbukkit.inventory.CraftShapedRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftShapedRecipe(result, this);
+        switch (this.recipeHeight) {
+            case 1:
+                switch (this.recipeWidth) {
+                    case 1:
+                        recipe.shape("a");
+                        break;
+                    case 2:
+                        recipe.shape("ab");
+                        break;
+                    case 3:
+                        recipe.shape("abc");
+                        break;
+                }
+                break;
+            case 2:
+                switch (this.recipeWidth) {
+                    case 1:
+                        recipe.shape("a","b");
+                        break;
+                    case 2:
+                        recipe.shape("ab","cd");
+                        break;
+                    case 3:
+                        recipe.shape("abc","def");
+                        break;
+                }
+                break;
+            case 3:
+                switch (this.recipeWidth) {
+                    case 1:
+                        recipe.shape("a","b","c");
+                        break;
+                    case 2:
+                        recipe.shape("ab","cd","ef");
+                        break;
+                    case 3:
+                        recipe.shape("abc","def","ghi");
+                        break;
+                }
+                break;
+        }
+        char c = 'a';
+        for (Ingredient list : this.recipeItems) {
+            if (list != null && list.matchingStacks.length > 0) {
+                net.minecraft.item.ItemStack stack = list.matchingStacks[0];
+                recipe.setIngredient(c, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.getItem()), (list.matchingStacks.length) > 1 ? 32767 : stack.getMetadata());
+            }
+            c++;
+        }
+        return recipe;
+    }
+
+    @Override
+    public void setKey(ResourceLocation key) {
+        this.key = key;
+    }
+    //Craftbukkit end
+
     //================================================ FORGE START ================================================
     @Override
     public int getRecipeWidth()
