--- ../src-base/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
@@ -17,6 +17,7 @@
     public final NonNullList<Ingredient> recipeItems;
     private final String group;
     private final boolean isSimple;
+    public net.minecraft.util.ResourceLocation key; //Craftbukkit TODO what is this for?
 
     public ShapelessRecipes(String group, ItemStack output, NonNullList<Ingredient> ingredients)
     {
@@ -136,4 +137,22 @@
     {
         return width * height >= this.recipeItems.size();
     }
+    //Craftbukkit methods
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.recipeOutput);
+        org.bukkit.craftbukkit.inventory.CraftShapelessRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftShapelessRecipe(result, this);
+        for (Ingredient list : this.recipeItems) {
+            if (list != null) {
+                net.minecraft.item.ItemStack stack = list.matchingStacks[0];
+                recipe.addIngredient(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.getItem()), (list.matchingStacks.length) > 1 ? 32767 : stack.getMetadata());
+            }
+        }
+        return recipe;
+    }
+
+    @Override
+    public void setKey(net.minecraft.util.ResourceLocation key) {
+        this.key = key;
+    }
 }
