--- ../src-base/minecraft/net/minecraft/network/NetworkManager.java
+++ ../src-work/minecraft/net/minecraft/network/NetworkManager.java
@@ -77,7 +77,11 @@
     private final Queue<NetworkManager.InboundHandlerTuplePacketListener> outboundPacketsQueue = Queues.<NetworkManager.InboundHandlerTuplePacketListener>newConcurrentLinkedQueue();
     private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
     private Channel channel;
-    private SocketAddress socketAddress;
+    // Spigot Start
+    public SocketAddress socketAddress; //private - public
+    public java.util.UUID spoofedUUID;
+    public com.mojang.authlib.properties.Property[] spoofedProfile;
+    // Spigot End
     private INetHandler packetListener;
     private ITextComponent terminationReason;
     private boolean isEncrypted;
@@ -300,7 +304,10 @@
     {
         if (this.channel.isOpen())
         {
+            /* Craftbukkit - We can't wait as this may be called from an event loop.
             this.channel.close().awaitUninterruptibly();
+            */
+            this.channel.close();
             this.terminationReason = message;
         }
     }
@@ -476,4 +483,5 @@
                 this.futureListeners = inFutureListeners;
             }
         }
+    public SocketAddress getRawAddress() { return this.channel.remoteAddress(); } //Spigot method
 }
