--- ../src-base/minecraft/net/minecraft/item/ItemEnderPearl.java
+++ ../src-work/minecraft/net/minecraft/item/ItemEnderPearl.java
@@ -35,7 +35,14 @@
         {
             EntityEnderPearl entityenderpearl = new EntityEnderPearl(worldIn, playerIn);
             entityenderpearl.shoot(playerIn, playerIn.rotationPitch, playerIn.rotationYaw, 0.0F, 1.5F, 1.0F);
-            worldIn.spawnEntity(entityenderpearl);
+            //Lava-test start - event, I don't care about the sound nor cooldown since the cooldown bar is still being shown on client.. TODO optional patch
+            //worldIn.spawnEntity(entityenderpearl);
+            if (!worldIn.spawnEntity(entityenderpearl)) {
+                if (!playerIn.capabilities.isCreativeMode) itemstack.grow(1); //revert
+                if (playerIn instanceof net.minecraft.entity.player.EntityPlayerMP) ((net.minecraft.entity.player.EntityPlayerMP) playerIn).getBukkitEntity().updateInventory();
+                return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+            }
+            //Lava-test end
         }
 
         playerIn.addStat(StatList.getObjectUseStats(this));
