--- ../src-base/minecraft/net/minecraft/item/ItemMap.java
+++ ../src-work/minecraft/net/minecraft/item/ItemMap.java
@@ -36,16 +36,24 @@
 
     public static ItemStack setupNewMap(World worldIn, double worldX, double worldZ, byte scale, boolean trackingPosition, boolean unlimitedTracking)
     {
-        ItemStack itemstack = new ItemStack(Items.FILLED_MAP, 1, worldIn.getUniqueDataId("map"));
+        //Craftbukkit start TODO check all this world messing, why is it there
+        World worldMain = worldIn.getServer().getServer().worlds[0]; // CraftBukkit - store reference to primary world
+        // ItemStack itemstack = new ItemStack(Items.FILLED_MAP, 1, worldIn.getUniqueDataId("map"));
+        ItemStack itemstack = new ItemStack(Items.FILLED_MAP, 1, worldMain.getUniqueDataId("map")); // CraftBukkit - use primary world for maps
+        //Craftbukkit end
         String s = "map_" + itemstack.getMetadata();
         MapData mapdata = new MapData(s);
-        worldIn.setData(s, mapdata);
+        worldMain.setData(s, mapdata);
         mapdata.scale = scale;
         mapdata.calculateMapCenter(worldX, worldZ, mapdata.scale);
-        mapdata.dimension = worldIn.provider.getDimension();
+        //Craftbukkit start
+        // mapdata.dimension = worldIn.provider.getDimension();
+        mapdata.dimension = ((net.minecraft.world.WorldServer) worldIn).dimension;  // CraftBukkit - use bukkit dimension
+        //Craftbukkit end
         mapdata.trackingPosition = trackingPosition;
         mapdata.unlimitedTracking = unlimitedTracking;
         mapdata.markDirty();
+        org.bukkit.craftbukkit.event.CraftEventFactory.callEvent(new org.bukkit.event.server.MapInitializeEvent(mapdata.mapView));
         return itemstack;
     }
 
@@ -60,19 +68,32 @@
     @Nullable
     public MapData getMapData(ItemStack stack, World worldIn)
     {
+        World worldMain = worldIn.getServer().getServer().worlds[0]; // CraftBukkit - store reference to primary world
         String s = "map_" + stack.getMetadata();
-        MapData mapdata = (MapData)worldIn.loadData(MapData.class, s);
+        //Craftbukkit start
+        // MapData mapdata = (MapData)worldIn.loadData(MapData.class, s);
+        MapData mapdata = (MapData) worldMain.loadData(MapData.class, s);// CraftBukkit - use primary world for maps
+        //Craftbukkit end
 
         if (mapdata == null && !worldIn.isRemote)
         {
-            stack.setItemDamage(worldIn.getUniqueDataId("map"));
+            stack.setItemDamage(worldMain.getUniqueDataId("map"));
             s = "map_" + stack.getMetadata();
             mapdata = new MapData(s);
             mapdata.scale = 3;
             mapdata.calculateMapCenter((double)worldIn.getWorldInfo().getSpawnX(), (double)worldIn.getWorldInfo().getSpawnZ(), mapdata.scale);
-            mapdata.dimension = worldIn.provider.getDimension();
+            //Craftbukkit start - fixes Bukkit multiworld maps
+            // mapdata.dimension = worldIn.provider.getDimension();
+            mapdata.dimension = ((net.minecraft.world.WorldServer) worldIn).dimension;
+            //Craftbukkit end
             mapdata.markDirty();
-            worldIn.setData(s, mapdata);
+            //Craftbukkit start - use primary world + event
+            //worldIn.setData(s, mapdata);
+            worldMain.setData(s, mapdata);
+
+            org.bukkit.event.server.MapInitializeEvent event = new org.bukkit.event.server.MapInitializeEvent(mapdata.mapView);
+            org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+            //Craftbukkit end
         }
 
         return mapdata;
@@ -80,7 +101,10 @@
 
     public void updateMapData(World worldIn, Entity viewer, MapData data)
     {
-        if (worldIn.provider.getDimension() == data.dimension && viewer instanceof EntityPlayer)
+        // CraftBukkit start - world.provider -> ((WorldServer) world) - this approach make it possible to update maps in other worlds prob TODO check this
+        //if (worldIn.provider.getDimension() == data.dimension && viewer instanceof EntityPlayer)
+        if (((net.minecraft.world.WorldServer) worldIn).dimension == data.dimension && viewer instanceof EntityPlayer)
+        //Craftbukkit end
         {
             int i = 1 << data.scale;
             int j = data.xCenter;
@@ -425,6 +449,7 @@
     protected static void scaleMap(ItemStack p_185063_0_, World p_185063_1_, int p_185063_2_)
     {
         MapData mapdata = Items.FILLED_MAP.getMapData(p_185063_0_, p_185063_1_);
+        p_185063_1_ = p_185063_1_.getServer().getServer().worlds[0]; //Craftbukkit - use primary world
         p_185063_0_.setItemDamage(p_185063_1_.getUniqueDataId("map"));
         MapData mapdata1 = new MapData("map_" + p_185063_0_.getMetadata());
 
@@ -436,12 +461,17 @@
             mapdata1.dimension = mapdata.dimension;
             mapdata1.markDirty();
             p_185063_1_.setData("map_" + p_185063_0_.getMetadata(), mapdata1);
+            //Craftbukkit start - event
+            org.bukkit.event.server.MapInitializeEvent event = new org.bukkit.event.server.MapInitializeEvent(mapdata1.mapView);
+            org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+            //Craftbukkit end
         }
     }
 
     protected static void enableMapTracking(ItemStack p_185064_0_, World p_185064_1_)
     {
         MapData mapdata = Items.FILLED_MAP.getMapData(p_185064_0_, p_185064_1_);
+        p_185064_1_ = p_185064_1_.getServer().getServer().worlds[0]; //Craftbukkit - use primary world
         p_185064_0_.setItemDamage(p_185064_1_.getUniqueDataId("map"));
         MapData mapdata1 = new MapData("map_" + p_185064_0_.getMetadata());
         mapdata1.trackingPosition = true;
@@ -454,6 +484,10 @@
             mapdata1.dimension = mapdata.dimension;
             mapdata1.markDirty();
             p_185064_1_.setData("map_" + p_185064_0_.getMetadata(), mapdata1);
+            //Craftbukkit start - event
+            org.bukkit.event.server.MapInitializeEvent event = new org.bukkit.event.server.MapInitializeEvent(mapdata1.mapView);
+            org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+            //Craftbukkit end
         }
     }
 
