--- ../src-base/minecraft/net/minecraft/item/ItemDye.java
+++ ../src-work/minecraft/net/minecraft/item/ItemDye.java
@@ -175,6 +175,12 @@
 
             if (!entitysheep.getSheared() && entitysheep.getFleeceColor() != enumdyecolor)
             {
+                //Craftbukkit start - event
+                org.bukkit.event.entity.SheepDyeWoolEvent event = new org.bukkit.event.entity.SheepDyeWoolEvent((org.bukkit.entity.Sheep) entitysheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData((byte)enumdyecolor.getMetadata()));
+                entitysheep.world.getServer().getPluginManager().callEvent(event);
+                if (event.isCancelled()) return false;
+                enumdyecolor = EnumDyeColor.byMetadata(event.getColor().getWoolData());
+                //Craftbukkit end
                 entitysheep.setFleeceColor(enumdyecolor);
                 stack.shrink(1);
             }
