--- ../src-base/minecraft/net/minecraft/item/ItemWrittenBook.java
+++ ../src-work/minecraft/net/minecraft/item/ItemWrittenBook.java
@@ -119,6 +119,24 @@
                         String s = nbttaglist.getStringTagAt(i);
                         ITextComponent itextcomponent;
 
+                        // CraftBukkit start tf is this for ??? TODO
+                        // Some commands use the worldserver variable but we leave it full of null values,
+                        // so we must temporarily populate it with the world of the commandsender
+                        net.minecraft.world.WorldServer[] prev = net.minecraft.server.MinecraftServer.getServerCB().worlds;
+                        net.minecraft.server.MinecraftServer server = net.minecraft.server.MinecraftServer.getServerCB();
+                        server.worlds = new net.minecraft.world.WorldServer[server.worldServerList.size()];
+                        server.worlds[0] = (net.minecraft.world.WorldServer) player.getEntityWorld();
+                        int bpos = 0;
+                        for (int pos = 1; pos < server.worlds.length; pos++) {
+                            net.minecraft.world.WorldServer world = server.worldServerList.get(bpos++);
+                            if (server.worlds[0] == world) {
+                                pos--;
+                                continue;
+                            }
+                            server.worlds[pos] = world;
+                        }
+                        // CraftBukkit end
+
                         try
                         {
                             itextcomponent = ITextComponent.Serializer.fromJsonLenient(s);
@@ -128,6 +146,7 @@
                         {
                             itextcomponent = new TextComponentString(s);
                         }
+                        finally { net.minecraft.server.MinecraftServer.getServerCB().worlds = prev; } //Craftbukkit
 
                         nbttaglist.set(i, new NBTTagString(ITextComponent.Serializer.componentToJson(itextcomponent)));
                     }
