--- ../src-base/minecraft/net/minecraft/item/ItemBoat.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBoat.java
@@ -84,6 +84,7 @@
             }
             else
             {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(playerIn, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, raytraceresult.getBlockPos(), raytraceresult.sideHit, itemstack, handIn).isCancelled()) return new ActionResult<>(EnumActionResult.PASS, itemstack); //Lava-test - event
                 Block block = worldIn.getBlockState(raytraceresult.getBlockPos()).getBlock();
                 boolean flag1 = block == Blocks.WATER || block == Blocks.FLOWING_WATER;
                 EntityBoat entityboat = new EntityBoat(worldIn, raytraceresult.hitVec.x, flag1 ? raytraceresult.hitVec.y - 0.12D : raytraceresult.hitVec.y, raytraceresult.hitVec.z);
@@ -98,7 +99,10 @@
                 {
                     if (!worldIn.isRemote)
                     {
-                        worldIn.spawnEntity(entityboat);
+                        //Craftbukkit start - make it cancelable if the boat won't spawn
+                        // worldIn.spawnEntity(entityboat);
+                        if (!worldIn.spawnEntity(entityboat)) return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                        //Craftbukkit end
                     }
 
                     if (!playerIn.capabilities.isCreativeMode)
