--- ../src-base/minecraft/net/minecraft/item/ItemStack.java
+++ ../src-work/minecraft/net/minecraft/item/ItemStack.java
@@ -56,7 +56,7 @@
     public static final DecimalFormat DECIMALFORMAT = new DecimalFormat("#.##");
     private int stackSize;
     private int animationsToGo;
-    private final Item item;
+    private Item item; //Lava?CB? - remove final for ability to set the item again
     private NBTTagCompound stackTagCompound;
     private boolean isEmpty;
     int itemDamage;
@@ -112,11 +112,28 @@
         this.forgeInit();
     }
 
+    public ItemStack(Item itemIn, int amount, int meta, boolean convert) {this(itemIn, amount, meta, null);} //Lava-test - ??? TODO
+
     private void updateEmptyState()
     {
         this.isEmpty = this.isEmpty();
     }
+    //Craftbukkit method Called to run this stack through the data converter to handle older storage methods and serialized items TODO ???
+    public void convertStack() {
+        if (net.minecraft.server.MinecraftServer.getServerCB() != null) {
+            // Don't convert beds - both the old and new data values are valid
+            // Conversion would make getting white beds (data value 0) impossible
+            if (this.item == Items.BED) {
+                return;
+            }
 
+            NBTTagCompound savedStack = new NBTTagCompound();
+            this.writeToNBT(savedStack);
+            net.minecraft.server.MinecraftServer.getServerCB().dataFixer.process(FixTypes.ITEM_INSTANCE, savedStack);
+            this.load(savedStack);
+        }
+    }
+    
     public ItemStack(NBTTagCompound compound)
     {
         this.capNBT = compound.hasKey("ForgeCaps") ? compound.getCompoundTag("ForgeCaps") : null;
@@ -138,6 +155,29 @@
         this.forgeInit();
     }
 
+    public void load(NBTTagCompound compound)
+    {
+        this.capNBT = compound.hasKey("ForgeCaps") ? compound.getCompoundTag("ForgeCaps") : null;
+        this.item = compound.hasKey("id", 8) ? Item.getByNameOrId(compound.getString("id")) : Items.AIR; //Forge fix tons of NumberFormatExceptions that are caused by deserializing EMPTY ItemStacks.
+        this.stackSize = compound.getByte("Count");
+        // CraftBukkit start - Route through setData for filtering
+        // this.itemDamage = Math.max(0, compound.getShort("Damage"));
+        this.setItemDamage(compound.getShort("Damage"));
+
+        if (compound.hasKey("tag", 10))
+        {
+            // CraftBukkit start - make defensive copy as this data may be coming from the save thread
+            // this.stackTagCompound = compound.getCompoundTag("tag");
+            this.stackTagCompound = compound.getCompoundTag("tag").copy();
+
+            if (this.item != null)
+            {
+                // this.item.updateItemStackNBT(compound);
+                this.item.updateItemStackNBT(this.stackTagCompound);
+            }
+        }
+    }
+
     public boolean isEmpty()
     {
         if (this == EMPTY)
@@ -189,7 +229,32 @@
         if (enumactionresult == EnumActionResult.SUCCESS)
         {
             playerIn.addStat(StatList.getObjectUseStats(this.item));
+                /*// Special case juke boxes as they update their tile entity. Copied from ItemRecord. TODO check this as forge doesn't have these special cases
+                if (this.item == Items.SKULL) { // Special case skulls to allow wither spawns to be cancelled
+                    BlockPos bp = pos;
+                    if (!worldIn.getBlockState(pos).getBlock().isReplaceable(worldIn, pos)) {
+                        if (!worldIn.getBlockState(pos).getMaterial().isSolid()) {
+                            bp = null;
+                        } else {
+                            bp = bp.offset(side);
+                        }
+                    }
+                    if (bp != null) {
+                        TileEntity te = worldIn.getTileEntity(bp);
+                        if (te instanceof TileEntitySkull) {
+                            Blocks.SKULL.checkWitherSpawn(worldIn, bp, (TileEntitySkull) te);
+                        }
+                    }
+                }
+
+                // SPIGOT-1288 - play sound stripped from ItemBlock ??? check TODO
+                if (this.item instanceof ItemBlock) {
+                    SoundType soundeffecttype = ((ItemBlock) this.item).getBlock().getSoundType();
+                    worldIn.playSound(playerIn, pos, soundeffecttype.getPlaceSound(), SoundCategory.BLOCKS, (soundeffecttype.getVolume() + 1.0F) / 2.0F, soundeffecttype.getPitch() * 0.8F);
+                }
+            }*/
         }
+        //worldIn.capturedTileEntities.clear(); ??? TODO what is this for, where is it used ...
 
         return enumactionresult;
     }
@@ -322,6 +387,19 @@
 
                 amount -= j;
 
+                // Spigot start
+                if (damager != null) {
+                    org.bukkit.craftbukkit.inventory.CraftItemStack item = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this);
+                    org.bukkit.event.player.PlayerItemDamageEvent event = new org.bukkit.event.player.PlayerItemDamageEvent(damager.getBukkitEntity(), item, amount);
+                    org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+                    if (amount != event.getDamage() || event.isCancelled()) {
+                        event.getPlayer().updateInventory();
+                    }
+                    if (event.isCancelled()) return false;
+                    amount = event.getDamage();
+                }
+                // Spigot end
+
                 if (amount <= 0)
                 {
                     return false;
@@ -347,6 +425,11 @@
                 if (this.attemptDamageItem(amount, entityIn.getRNG(), entityIn instanceof EntityPlayerMP ? (EntityPlayerMP)entityIn : null))
                 {
                     entityIn.renderBrokenItemStack(this);
+                    // CraftBukkit start - Check for item breaking
+                    if (this.stackSize == 1 && entityIn instanceof EntityPlayer) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemBreakEvent((EntityPlayer) entityIn, this);
+                    }
+                    // CraftBukkit end
                     this.shrink(1);
 
                     if (entityIn instanceof EntityPlayer)
@@ -1043,6 +1126,13 @@
         nbttaglist.appendTag(nbttagcompound);
     }
 
+    @Deprecated //Lava?CB? add method for ability to set item
+    public void setItem(Item item) {
+        this.item = item;
+        // Update delegate as well
+        this.delegate = item.delegate;
+    }
+
     public ITextComponent getTextComponent()
     {
         TextComponentString textcomponentstring = new TextComponentString(this.getDisplayName());
