--- ../src-base/minecraft/net/minecraft/item/ItemShears.java
+++ ../src-work/minecraft/net/minecraft/item/ItemShears.java
@@ -50,6 +50,13 @@
             BlockPos pos = new BlockPos(entity.posX, entity.posY, entity.posZ);
             if (target.isShearable(itemstack, entity.world, pos))
             {
+                // CraftBukkit start
+                org.bukkit.event.player.PlayerShearEntityEvent event = new org.bukkit.event.player.PlayerShearEntityEvent((org.bukkit.entity.Player) player.getBukkitEntity(), entity.getBukkitEntity());
+                player.world.getServer().getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    return false;
+                }
+                // CraftBukkit end
                 java.util.List<ItemStack> drops = target.onSheared(itemstack, entity.world, pos,
                         net.minecraft.enchantment.EnchantmentHelper.getEnchantmentLevel(net.minecraft.init.Enchantments.FORTUNE, itemstack));
 
