--- ../src-base/minecraft/net/minecraft/item/ItemBucket.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBucket.java
@@ -72,6 +72,7 @@
 
                     if (material == Material.WATER && ((Integer)iblockstate.getValue(BlockLiquid.LEVEL)).intValue() == 0)
                     {
+                        if (org.lavapowered.lava.events.LavaEventFactory.onBucketFill(playerIn, itemstack, blockpos, true, this)) return new ActionResult<>(EnumActionResult.FAIL, itemstack); //Lava-test event
                         worldIn.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 11);
                         playerIn.addStat(StatList.getObjectUseStats(this));
                         playerIn.playSound(SoundEvents.ITEM_BUCKET_FILL, 1.0F, 1.0F);
@@ -79,6 +80,7 @@
                     }
                     else if (material == Material.LAVA && ((Integer)iblockstate.getValue(BlockLiquid.LEVEL)).intValue() == 0)
                     {
+                        if (org.lavapowered.lava.events.LavaEventFactory.onBucketFill(playerIn, itemstack, blockpos, false, this)) return new ActionResult<>(EnumActionResult.FAIL, itemstack); //Lava-test - event
                         playerIn.playSound(SoundEvents.ITEM_BUCKET_FILL_LAVA, 1.0F, 1.0F);
                         worldIn.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 11);
                         playerIn.addStat(StatList.getObjectUseStats(this));
@@ -94,6 +96,10 @@
             {
                 boolean flag1 = worldIn.getBlockState(blockpos).getBlock().isReplaceable(worldIn, blockpos);
                 BlockPos blockpos1 = flag1 && raytraceresult.sideHit == EnumFacing.UP ? blockpos : blockpos.offset(raytraceresult.sideHit);
+                //Lava-test start - add fields instead of rewriting method args
+                this.eventFacing = raytraceresult.sideHit;
+                this.eventStack = itemstack;
+                //Lava-test end
 
                 if (!playerIn.canPlayerEdit(blockpos1, raytraceresult.sideHit, itemstack))
                 {
@@ -119,6 +125,7 @@
 
     private ItemStack fillBucket(ItemStack emptyBuckets, EntityPlayer player, Item fullBucket)
     {
+        if (this.eventStack != null) fullBucket = this.eventStack.getItem(); //Lava-test - move the event result here TODO support for whole ItemStack? (nbt, amount...)
         if (player.capabilities.isCreativeMode)
         {
             return emptyBuckets;
@@ -162,6 +169,7 @@
             }
             else
             {
+                if (player != null && org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerBucketEmptyEvent(player, posIn.getX(), posIn.getY(), posIn.getZ(), this.eventFacing, this.eventStack).isCancelled()) return false; //Lava-test - event
                 if (worldIn.provider.doesWaterVaporize() && this.containedBlock == Blocks.FLOWING_WATER)
                 {
                     int l = posIn.getX();
@@ -202,4 +210,13 @@
             return super.initCapabilities(stack, nbt);
         }
     }
+    //Lava-test start - fields for events and CB method, just in case some bad plugins use it
+    private EnumFacing eventFacing;
+    public ItemStack eventStack;
+    public boolean tryPlaceContainedLiquid(@Nullable EntityPlayer player, World worldIn, BlockPos posIn, EnumFacing enumdirection, BlockPos clicked, ItemStack itemstack) {
+        this.eventFacing = enumdirection;
+        this.eventStack = itemstack;
+        return this.tryPlaceContainedLiquid(player, worldIn, posIn);
+    }
+    //Lava-test end
 }
