--- ../src-base/minecraft/net/minecraft/item/ItemFishingRod.java
+++ ../src-work/minecraft/net/minecraft/item/ItemFishingRod.java
@@ -74,7 +74,7 @@
         }
         else
         {
-            worldIn.playSound((EntityPlayer)null, playerIn.posX, playerIn.posY, playerIn.posZ, SoundEvents.ENTITY_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (itemRand.nextFloat() * 0.4F + 0.8F));
+            // worldIn.playSound((EntityPlayer)null, playerIn.posX, playerIn.posY, playerIn.posZ, SoundEvents.ENTITY_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (itemRand.nextFloat() * 0.4F + 0.8F)); // CraftBukkit - moved down
 
             if (!worldIn.isRemote)
             {
@@ -92,7 +92,17 @@
                 {
                     entityfishhook.setLuck(k);
                 }
+                //Craftbukkit start - event
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) playerIn.getBukkitEntity(), null, (org.bukkit.entity.Fish) entityfishhook.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.FISHING);
+                worldIn.getServer().getPluginManager().callEvent(playerFishEvent);
 
+                if (playerFishEvent.isCancelled()) {
+                    playerIn.fishEntity = null;
+                    return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                }
+
+                worldIn.playSound(null, playerIn.posX, playerIn.posY, playerIn.posZ, SoundEvents.ENTITY_BOBBER_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (itemRand.nextFloat() * 0.4F + 0.8F));
+                //Craftbukkit end
                 worldIn.spawnEntity(entityfishhook);
             }
 
