--- ../src-base/minecraft/net/minecraft/item/ItemHangingEntity.java
+++ ../src-work/minecraft/net/minecraft/item/ItemHangingEntity.java
@@ -35,6 +35,16 @@
             {
                 if (!worldIn.isRemote)
                 {
+                    //Craftbukkit start - event
+                    org.bukkit.entity.Player who = (player == null) ? null : (org.bukkit.entity.Player) player.getBukkitEntity();
+                    org.bukkit.block.Block blockClicked = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                    org.bukkit.block.BlockFace blockFace = org.bukkit.craftbukkit.block.CraftBlock.notchToBlockFace(facing);
+
+                    org.bukkit.event.hanging.HangingPlaceEvent event = new org.bukkit.event.hanging.HangingPlaceEvent((org.bukkit.entity.Hanging) entityhanging.getBukkitEntity(), who, blockClicked, blockFace);
+                    worldIn.getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) return EnumActionResult.FAIL;
+                    //Craftbukkit end
                     entityhanging.playPlaceSound();
                     worldIn.spawnEntity(entityhanging);
                 }
