--- ../src-base/minecraft/net/minecraft/command/CommandSpreadPlayers.java
+++ ../src-work/minecraft/net/minecraft/command/CommandSpreadPlayers.java
@@ -388,5 +388,12 @@
                 this.x = MathHelper.nextDouble(rand, p_111097_2_, p_111097_6_);
                 this.z = MathHelper.nextDouble(rand, p_111097_4_, p_111097_8_);
             }
+
+            // CraftBukkit start - add a version of getBlockState which force loads chunks /// Lava-test - investigate why is this here TODO
+            /*private static IBlockState getBlockState(World world, BlockPos position) {
+                ((ChunkProviderServer) world.getChunkProvider()).loadChunk(position.getX() >> 4, position.getZ() >> 4);
+                return world.getBlockState(position);
+            }*/
+            // CraftBukkit end
         }
 }
