--- ../src-base/minecraft/net/minecraft/command/CommandExecuteAt.java
+++ ../src-work/minecraft/net/minecraft/command/CommandExecuteAt.java
@@ -78,7 +78,11 @@
 
             try
             {
-                int j = icommandmanager.executeCommand(icommandsender, s);
+                //Craftbukkit start - ??? TODO investigate
+                //int j = icommandmanager.executeCommand(icommandsender, s);
+                org.bukkit.command.CommandSender bukkitSender = net.minecraft.tileentity.CommandBlockBaseLogic.unwrapSender(sender);
+                int j = net.minecraft.tileentity.CommandBlockBaseLogic.executeCommand(icommandsender, new org.bukkit.craftbukkit.command.ProxiedNativeCommandSender(icommandsender, bukkitSender, entity.getBukkitEntity()), s);
+                //Craftbukkit end
 
                 if (j < 1)
                 {
@@ -87,6 +91,11 @@
             }
             catch (Throwable var23)
             {
+                //Craftbukkit start - ??? TODO investigate
+                /*if (var23 instanceof CommandException) {
+                    throw (CommandException) var23;
+                }*/
+                //Craftbukkit end
                 throw new CommandException("commands.execute.failed", new Object[] {s, entity.getName()});
             }
         }
