--- ../src-base/minecraft/net/minecraft/command/CommandDebug.java
+++ ../src-work/minecraft/net/minecraft/command/CommandDebug.java
@@ -40,6 +40,14 @@
 
     public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException
     {
+        // CraftBukkit start - only allow use when enabled (so that no blank profile results occur)
+        if (!server.profiler.ENABLED) {
+            sender.sendMessage(new net.minecraft.util.text.TextComponentString("Vanilla debug profiling is disabled."));
+            sender.sendMessage(new net.minecraft.util.text.TextComponentString("To enable, restart the server with `-DenableDebugMethodProfiler=true' before `-jar'."));
+            sender.sendMessage(new net.minecraft.util.text.TextComponentString("Use `/timings' for plugin timings."));
+            return;
+        }
+        // CraftBukkit end
         if (args.length < 1)
         {
             throw new WrongUsageException("commands.debug.usage", new Object[0]);
