--- ../src-base/minecraft/net/minecraft/command/FunctionObject.java
+++ ../src-work/minecraft/net/minecraft/command/FunctionObject.java
@@ -108,7 +108,16 @@
 
             public void execute(FunctionManager functionManagerIn, ICommandSender sender, ArrayDeque<FunctionManager.QueuedCommand> commandQueue, int maxCommandChainLength)
             {
-                functionManagerIn.getCommandManager().executeCommand(sender, this.command);
+                //Craftbukkit start - ??? TODO investigate
+//                functionManagerIn.getCommandManager().executeCommand(sender, this.command);
+                org.bukkit.command.CommandSender bukkitSender;
+                if (sender instanceof FunctionManager.CustomFunctionListener) {
+                    bukkitSender = ((FunctionManager.CustomFunctionListener) sender).sender;
+                } else {
+                    bukkitSender = net.minecraft.tileentity.CommandBlockBaseLogic.unwrapSender(sender);
+                }
+                net.minecraft.tileentity.CommandBlockBaseLogic.executeSafely(sender, bukkitSender, this.command);
+                //Craftbukkit end
             }
 
             public String toString()
