--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketChat.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketChat.java
@@ -9,10 +9,13 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
+import java.io.IOException;
+
 public class SPacketChat implements Packet<INetHandlerPlayClient>
 {
     private ITextComponent chatComponent;
     private ChatType type;
+    public net.md_5.bungee.api.chat.BaseComponent[] components; // Spigot
 
     public SPacketChat()
     {
@@ -23,9 +26,14 @@
         this(componentIn, ChatType.SYSTEM);
     }
 
-    public SPacketChat(ITextComponent message, ChatType type)
+    public SPacketChat(ITextComponent message, byte type)
     {
         this.chatComponent = message;
+        this.type = ChatType.byId(type);
+    }
+    //Spigot constructor - ??? not used anywhere TODO check
+    public SPacketChat(ITextComponent message, ChatType type) {
+        this.chatComponent = message;
         this.type = type;
     }
 
@@ -35,8 +43,12 @@
         this.type = ChatType.byId(buf.readByte());
     }
 
-    public void writePacketData(PacketBuffer buf) throws IOException
-    {
+    public void writePacketData(PacketBuffer buf) throws IOException {
+        // Spigot start
+        //buf.writeTextComponent(this.chatComponent);
+        if (components != null) buf.writeString(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+        else
+        // Spigot end
         buf.writeTextComponent(this.chatComponent);
         buf.writeByte(this.type.getId());
     }
