--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketWorldBorder.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketWorldBorder.java
@@ -10,7 +10,7 @@
 
 public class SPacketWorldBorder implements Packet<INetHandlerPlayClient>
 {
-    private SPacketWorldBorder.Action action;
+    private Action action;
     private int size;
     private double centerX;
     private double centerZ;
@@ -27,8 +27,12 @@
     public SPacketWorldBorder(WorldBorder border, SPacketWorldBorder.Action actionIn)
     {
         this.action = actionIn;
-        this.centerX = border.getCenterX();
-        this.centerZ = border.getCenterZ();
+        // CraftBukkit start - multiply out nether border
+        // this.centerX = border.getCenterX();
+        // this.centerZ = border.getCenterZ();
+        this.centerX = border.getCenterX() * (border.world.provider instanceof net.minecraft.world.WorldProviderHell ? 8 : 1);
+        this.centerZ = border.getCenterZ() * (border.world.provider instanceof net.minecraft.world.WorldProviderHell ? 8 : 1);
+        // CraftBukkit end
         this.diameter = border.getDiameter();
         this.targetSize = border.getTargetSize();
         this.timeUntilTarget = border.getTimeUntilTarget();
