--- ../src-base/minecraft/net/minecraftforge/common/WorldSpecificSaveHandler.java
+++ ../src-work/minecraft/net/minecraftforge/common/WorldSpecificSaveHandler.java
@@ -21,7 +21,7 @@
 
 import java.io.File;
 import java.io.IOException;
-
+import java.util.UUID;
 import net.minecraft.world.gen.structure.template.TemplateManager;
 import org.apache.logging.log4j.Level;
 
@@ -84,26 +84,25 @@
         return file;
     }
 
-    private void copyFile(String name, File to)
-    {
+    private void copyFile(String name, File to) {
         File parentFile = parent.getMapFileFromName(name);
-        if (parentFile.exists())
-        {
-            try
-            {
+        if (parentFile.exists()) {
+            try {
                 Files.copy(parentFile, to);
-            }
-            catch (IOException e)
-            {
+            } catch (IOException e) {
                 FMLLog.log.error("A critical error occurred copying {} to world specific dat folder - new file will be created.", parentFile.getName(), e);
             }
         }
     }
 
     @Override
-    public TemplateManager getStructureTemplateManager()
-    {
+    public TemplateManager getStructureTemplateManager() {
         return parent.getStructureTemplateManager();
     }
 
+    @Override
+    public UUID getUUID() {
+        return parent.getUUID();
+    }
+
 }
